
<!-- saved from url=(0113)https://b.0-0.plus/blog/md.htm?src=https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS.md -->
<html lang="zh-cn"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <script type="module" src="./mofa_files/zero-md.min.js"></script>
    
<script src="./mofa_files/marked.min.js"></script><script src="./mofa_files/prism.min.js" data-manual=""></script><script src="./mofa_files/prism-autoloader.min.js"></script></head>

<body>
    <div style="width:70%; margin:0 auto;" id="mdcontainer">

    <zero-md src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS.md" no-shadow=""><template shadowrootmode="open"><div class="markdown-styles" data-hash="ogj29p"><style>:host{display:block;position:relative;contain:content;}:host([hidden]){display:none;}</style><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/sindresorhus/github-markdown-css@4/github-markdown.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/PrismJS/prism@1/themes/prism.min.css"></div><div class="markdown-body" data-hash="15sspbz"><h2 id="mofa-cosmos-pipeline">MOFA-COSMOS pipeline</h2><h3 id="installation-and-dependency">Installation and dependency</h3><p>In this tutorial, we use the new Meta-Footprint Analysis MOON to score a
mechanistic network.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#install cosmosR from github directly to obtain the newest version</span>
<span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"devtools"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> install.packages<span class="token punctuation">(</span><span class="token string">"devtools"</span><span class="token punctuation">)</span>
<span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"cosmosR"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> devtools<span class="token operator">::</span>install_github<span class="token punctuation">(</span><span class="token string">"saezlab/cosmosR"</span><span class="token punctuation">)</span>
</code></pre><p>DecoupleR is used to score various regulatory interactions.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> install.packages<span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">)</span>
<span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"decoupleR"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> BiocManager<span class="token operator">::</span>install<span class="token punctuation">(</span><span class="token string">"saezlab/decoupleR"</span><span class="token punctuation">)</span>
</code></pre><p>We are using the LIANA package to process the LR interactions.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"remotes"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> install.packages<span class="token punctuation">(</span><span class="token string">"remotes"</span><span class="token punctuation">)</span>
<span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"liana"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> remotes<span class="token operator">::</span>install_github<span class="token punctuation">(</span><span class="token string">'saezlab/liana'</span><span class="token punctuation">)</span>
</code></pre><p>We are using MOFA2 (Argelaguet et al., 2018) to find decompose variance
across different omics and use COSMOS to find mechanistic
interpretations of its factors. However, since we are using the python
version of MOFA2, please make sure to have
<a href="https://github.com/bioFAM/mofapy2">mofapy2</a> as well as panda and numpy
installed in your (conda) environment (e.g.&nbsp;by using reticulate:
reticulate::use_condaenv(“base”, required = T) %&gt;%
reticulate::conda_install(c(“mofapy2”,“panda”,“numpy”)). For downstream
analysis, the R version of MOFA2 is used.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Install MOFA2 from bioconductor (stable version)</span>
<span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> install.packages<span class="token punctuation">(</span><span class="token string">"BiocManager"</span><span class="token punctuation">)</span>
<span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">"MOFA2"</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span> BiocManager<span class="token operator">::</span>install<span class="token punctuation">(</span><span class="token string">"MOFA2"</span><span class="token punctuation">)</span>
</code></pre><p>General information (tutorials) on how to use COSMOS and MOFA2, see
<a href="https://saezlab.github.io/cosmosR/articles/tutorial.html">COSMOS</a> and
<a href="https://biofam.github.io/MOFA2/tutorials.html">MOFA2</a>.</p><p>For data manipulation, visualization, etc. further packages are needed
which are loaded here along with MOFA2 and COSMOS.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#imports</span>
library<span class="token punctuation">(</span>readr<span class="token punctuation">)</span>

<span class="token comment">#core methods</span>
library<span class="token punctuation">(</span>cosmosR<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>MOFA2<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>liana<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>decoupleR<span class="token punctuation">)</span>

<span class="token comment">#data manipulations</span>
library<span class="token punctuation">(</span>dplyr<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>reshape2<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>GSEABase<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>tidyr<span class="token punctuation">)</span>

<span class="token comment">#plotting</span>
library<span class="token punctuation">(</span>ggplot2<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>ggfortify<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>pheatmap<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>gridExtra<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>RColorBrewer<span class="token punctuation">)</span>
<span class="token comment"># library(RCy3)</span>
</code></pre><h3 id="from-omics-data-to-mofa-ready-input">From omics data to MOFA ready input</h3><p>Since extensive pre-processing is beyond the scope of this tutorial,
here only the transformation of pre-processed single omics data to a
long data.frame (columns: “sample”, “group”, “feature”, “view”, “value”)
is shown. For more details and information regarding appropriate input,
please refer to the MOFA tutorial <a href="https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/getting_started_R.html">MOFA2: training a model in
R</a>.
The raw data is accessed through <a href="https://discover.nci.nih.gov/cellminer/home.do">NCI60
cellminer</a> and the
scripts for pre-processing can be found in the associated folder.</p><p>Here, we have the following views (with samples as columns and genes as
rows): transcriptomics (RNA-seq
<a href="https://pubmed.ncbi.nlm.nih.gov/31113817/"></a><a href="pmid:31113817">PMID:31113817</a>),
proteomics (SWATH (Mass spectrometry)
<a href="https://pubmed.ncbi.nlm.nih.gov/31733513/%3E"></a><a href="pmid:31733513">PMID:31733513</a>} and
metabolomics (LC/MS &amp; GC/MS (Mass spectrometry) <a href="https://wiki.nci.nih.gov/display/NCIDTPdata/Molecular+Target+Data">DTP NCI60
data</a>).
The group information is stored inside the RNA metadata and based on
transcription factor clustering (see scripts/RNA/Analyze_RNA.R).
Further, only samples are kept for which each omic is available (however
this is not necessary since MOFA2 is able to impute data).</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Transcriptomics</span>
<span class="token comment">## Load data</span>
RNA <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>read_csv<span class="token punctuation">(</span><span class="token string">"data/RNA/RNA_log2_FPKM_clean.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
rownames<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>
RNA <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

<span class="token comment">## Remove genes with excessive amount of NAs (only keep genes with max. amount </span>
<span class="token comment">#of NAs = 33.3 % across cell lines)</span>
RNA <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span>rowSums<span class="token punctuation">(</span>is.na<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token operator">&lt;</span><span class="token punctuation">(</span>dim<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token operator">/</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>

<span class="token comment">## Only keep highly variable genes (as suggested by MOFA): here top ≈70% genes </span>
<span class="token comment">#to keep &gt;75% of TFs (103 out of 133). For stronger selection, we can further </span>
<span class="token comment">#reduce the number of genes (e.g. top 50%)</span>
RNA_sd <span class="token operator">&lt;-</span> sort<span class="token punctuation">(</span>apply<span class="token punctuation">(</span>RNA<span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span> sd<span class="token punctuation">(</span>x<span class="token punctuation">,</span>na.rm <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span>
hist<span class="token punctuation">(</span>RNA_sd<span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Preparing%20MOFA%20input-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">hist<span class="token punctuation">(</span>RNA_sd<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">6000</span><span class="token punctuation">]</span><span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Preparing%20MOFA%20input-2.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">RNA_sd <span class="token operator">&lt;-</span> RNA_sd<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">6000</span><span class="token punctuation">]</span>
RNA <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span>names<span class="token punctuation">(</span>RNA_sd<span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>

<span class="token comment"># Proteomics</span>
<span class="token comment">## Load data</span>
proteo <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>read_csv<span class="token punctuation">(</span><span class="token string">"data/proteomic/Prot_log10_SWATH_clean.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
rownames<span class="token punctuation">(</span>proteo<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> proteo<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>
proteo <span class="token operator">&lt;-</span> proteo<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

<span class="token comment">## Only keep highly variable proteins (as suggested by MOFA): here top ≈60%. </span>
<span class="token comment">#For stronger selection, we can further reduce the number of proteins (e.g. only keep top 50%)</span>
proteo_sd <span class="token operator">&lt;-</span> sort<span class="token punctuation">(</span>apply<span class="token punctuation">(</span>proteo<span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span> sd<span class="token punctuation">(</span>x<span class="token punctuation">,</span>na.rm <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span>
hist<span class="token punctuation">(</span>proteo_sd<span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Preparing%20MOFA%20input-3.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">hist<span class="token punctuation">(</span>proteo_sd<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span>round<span class="token punctuation">(</span>dim<span class="token punctuation">(</span>proteo<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token operator">*</span><span class="token number">3</span><span class="token operator">/</span><span class="token number">5</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Preparing%20MOFA%20input-4.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">proteo_sd <span class="token operator">&lt;-</span> proteo_sd<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span>round<span class="token punctuation">(</span>dim<span class="token punctuation">(</span>proteo<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token operator">*</span><span class="token number">3</span><span class="token operator">/</span><span class="token number">5</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
proteo <span class="token operator">&lt;-</span> proteo<span class="token punctuation">[</span>names<span class="token punctuation">(</span>proteo_sd<span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>

<span class="token comment"># Metabolomics</span>
<span class="token comment">## Load data</span>
metab <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>read_csv<span class="token punctuation">(</span><span class="token string">"data/metabolomic/metabolomic_clean_vsn.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
rownames<span class="token punctuation">(</span>metab<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> metab<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>
metab <span class="token operator">&lt;-</span> metab<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

<span class="token comment">## Since we have only a limited number of metabolite measurements, </span>
<span class="token comment">#all measurements are kept here</span>
metab_sd <span class="token operator">&lt;-</span> apply<span class="token punctuation">(</span>metab<span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span> sd<span class="token punctuation">(</span>x<span class="token punctuation">,</span>na.rm<span class="token operator">=</span>T<span class="token punctuation">)</span><span class="token punctuation">)</span>
hist<span class="token punctuation">(</span>metab_sd<span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Preparing%20MOFA%20input-5.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Create long data frame</span>
<span class="token comment">## Only keep samples with each view present </span>
overlap_patients <span class="token operator">&lt;-</span> intersect<span class="token punctuation">(</span>intersect<span class="token punctuation">(</span>names<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>proteo<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>metab<span class="token punctuation">)</span><span class="token punctuation">)</span>
RNA <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span><span class="token punctuation">,</span>overlap_patients<span class="token punctuation">]</span>
proteo <span class="token operator">&lt;-</span> proteo<span class="token punctuation">[</span><span class="token punctuation">,</span>overlap_patients<span class="token punctuation">]</span>
metab <span class="token operator">&lt;-</span> metab<span class="token punctuation">[</span><span class="token punctuation">,</span>overlap_patients<span class="token punctuation">]</span>

<span class="token comment">## Create columns required for MOFA</span>
<span class="token comment">### RNA</span>
RNA <span class="token operator">&lt;-</span> melt<span class="token punctuation">(</span>as.data.frame<span class="token punctuation">(</span>cbind<span class="token punctuation">(</span>RNA<span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
RNA<span class="token operator">$</span>view <span class="token operator">&lt;-</span> <span class="token string">"RNA"</span>
RNA <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>                 
names<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"sample"</span><span class="token punctuation">,</span><span class="token string">"feature"</span><span class="token punctuation">,</span><span class="token string">"view"</span><span class="token punctuation">,</span><span class="token string">"value"</span><span class="token punctuation">)</span>

<span class="token comment">### Proteomics</span>
proteo <span class="token operator">&lt;-</span> melt<span class="token punctuation">(</span>as.data.frame<span class="token punctuation">(</span>cbind<span class="token punctuation">(</span>proteo<span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>proteo<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
proteo<span class="token operator">$</span>view <span class="token operator">&lt;-</span> <span class="token string">"proteo"</span>
proteo <span class="token operator">&lt;-</span> proteo<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>                 
names<span class="token punctuation">(</span>proteo<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"sample"</span><span class="token punctuation">,</span><span class="token string">"feature"</span><span class="token punctuation">,</span><span class="token string">"view"</span><span class="token punctuation">,</span><span class="token string">"value"</span><span class="token punctuation">)</span>

<span class="token comment">### Metabolomics</span>
metab <span class="token operator">&lt;-</span> melt<span class="token punctuation">(</span>as.data.frame<span class="token punctuation">(</span>cbind<span class="token punctuation">(</span>metab<span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>metab<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
metab<span class="token operator">$</span>view <span class="token operator">&lt;-</span> <span class="token string">"metab"</span>
metab <span class="token operator">&lt;-</span> metab<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>                 
names<span class="token punctuation">(</span>metab<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"sample"</span><span class="token punctuation">,</span><span class="token string">"feature"</span><span class="token punctuation">,</span><span class="token string">"view"</span><span class="token punctuation">,</span><span class="token string">"value"</span><span class="token punctuation">)</span>

<span class="token comment">## Merge long data frame to one</span>
mofa_ready_data <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>do.call<span class="token punctuation">(</span>rbind<span class="token punctuation">,</span>list<span class="token punctuation">(</span>RNA<span class="token punctuation">,</span>proteo<span class="token punctuation">,</span>metab<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment"># Save MOFA metadata information and MOFA input</span>
write_csv<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"data/mofa/mofa_data.csv"</span><span class="token punctuation">)</span>
</code></pre><p>The final data frame has the following structure:</p><pre class="language-r" tabindex="0"><code class="language-r">head<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">##   sample  feature view  value</span>
<span class="token comment">## 1  786-0     KRT8  RNA  7.261</span>
<span class="token comment">## 2  786-0      FN1  RNA  7.823</span>
<span class="token comment">## 3  786-0   S100A4  RNA     NA</span>
<span class="token comment">## 4  786-0 RNA5-8S5  RNA 11.045</span>
<span class="token comment">## 5  786-0     MT1E  RNA  7.010</span>
<span class="token comment">## 6  786-0      VIM  RNA  9.366</span>
</code></pre><p>We have successfully combined single omic data sets to a long
multi-omics data frame that can be used as a MOFA input.</p><p>In this part we assess the correlation between the RNA and proteomic
data. We both look for correlation across samples AND across features.
When we check across sample, we interrogate the correlation at the level
of single genes, that is for given gene, how is a change is transcript
abundance is reflected in the the level of its protein. This correlation
can be different between genes, as there are all subject to different
regulatory mechanisms. when we check across features, we interrogate the
correlation at the level of single samples. that is, we check if overall
the protein abundance are good proxy of transcript abundance, and
vice-versa. The imperfect correlation between trnaxcripts and proteins
reflects the complexity of the translation process, as well as protein
and RNA stability.</p><pre class="language-r" tabindex="0"><code class="language-r">condition_prot_RNA_correlation <span class="token operator">&lt;-</span> sapply<span class="token punctuation">(</span>unique<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>sample<span class="token punctuation">)</span><span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>condition<span class="token punctuation">)</span><span class="token punctuation">{</span>
  merged_data <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>sample <span class="token operator">==</span> condition <span class="token operator">&amp;</span> mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"RNA"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span>
                       mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>sample <span class="token operator">==</span> condition <span class="token operator">&amp;</span> mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"proteo"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span>
                       by <span class="token operator">=</span> <span class="token string">"feature"</span><span class="token punctuation">)</span>
  return<span class="token punctuation">(</span>cor<span class="token punctuation">(</span>merged_data<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">,</span>merged_data<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">,</span> use <span class="token operator">=</span> <span class="token string">"pairwise.complete.obs"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

hist<span class="token punctuation">(</span>condition_prot_RNA_correlation<span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">20</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Supp%20figure%20condition_prot_RNA_correlation-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">overlap_genes <span class="token operator">&lt;-</span> unique<span class="token punctuation">(</span>intersect<span class="token punctuation">(</span>proteo<span class="token operator">$</span>feature<span class="token punctuation">,</span>RNA<span class="token operator">$</span>feature<span class="token punctuation">)</span><span class="token punctuation">)</span>
single_prot_RNA_correlation <span class="token operator">&lt;-</span> sapply<span class="token punctuation">(</span>overlap_genes<span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>gene<span class="token punctuation">)</span><span class="token punctuation">{</span>
  merged_data <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>feature <span class="token operator">==</span> gene <span class="token operator">&amp;</span> mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"RNA"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span>
                       mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>feature <span class="token operator">==</span> gene <span class="token operator">&amp;</span> mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"proteo"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span>
                       by <span class="token operator">=</span> <span class="token string">"sample"</span><span class="token punctuation">)</span>
  return<span class="token punctuation">(</span>cor<span class="token punctuation">(</span>merged_data<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">,</span>merged_data<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">,</span> use <span class="token operator">=</span> <span class="token string">"pairwise.complete.obs"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

hist<span class="token punctuation">(</span>single_prot_RNA_correlation<span class="token punctuation">,</span> breaks <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Supp%20figure%20single_prot_RNA_correlation-1.png" alt=""><!-- --></p><h3 id="mofa-run">MOFA run</h3><p>After pre-processing the data into MOFA appropriate input, let’s perform
the MOFA analysis. To change specific MOFA options the function
write_MOFA_options_file is used. For more information regarding option
setting see <a href="https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/getting_started_R.html">MOFA option
tutorial</a>.
Since we don’t know how many factors are needed to explain our data
well, various maximum numbers of factors are tested. The final results
can be found in the results/mofa/ folder.</p><p>Only run this chunk if you want to re-perform the mofa optimisation. The
results are already available in the results/mofa folder. We are testing
various amount of maximum number of factors that MOFA is allowed to
explore.</p><p>You need to instal mofapy2 in python3 to run this chunk. On windows, you
can simply run “python3” the terminal to make sure its installed or be
automatically redirected to the windows store page if it’s not. Then,
you can run “python3 -m pip install –user mofapy2” in a terminal to
install the module to python3.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token keyword">for</span><span class="token punctuation">(</span>i <span class="token keyword">in</span> c<span class="token punctuation">(</span><span class="token number">5</span><span class="token operator">:</span><span class="token number">15</span><span class="token punctuation">,</span><span class="token number">20</span><span class="token punctuation">,</span>length<span class="token punctuation">(</span>unique<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>sample<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">{</span>
wd <span class="token operator">&lt;-</span> getwd<span class="token punctuation">(</span><span class="token punctuation">)</span>
write_MOFA_options_file<span class="token punctuation">(</span>input_file <span class="token operator">=</span> <span class="token string">'/data/mofa/mofa_data.csv'</span><span class="token punctuation">,</span> output_file <span class="token operator">=</span> paste0<span class="token punctuation">(</span><span class="token string">'/results/mofa/mofa_res_'</span><span class="token punctuation">,</span>i<span class="token punctuation">,</span><span class="token string">'factor.hdf5'</span><span class="token punctuation">)</span><span class="token punctuation">,</span> factors <span class="token operator">=</span> i<span class="token punctuation">,</span> convergence_mode <span class="token operator">=</span> <span class="token string">"slow"</span><span class="token punctuation">,</span> likelihoods <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">'gaussian'</span><span class="token punctuation">,</span><span class="token string">'gaussian'</span><span class="token punctuation">,</span><span class="token string">'gaussian'</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
system<span class="token punctuation">(</span>paste0<span class="token punctuation">(</span>paste<span class="token punctuation">(</span><span class="token string">'python3'</span><span class="token punctuation">,</span> wd<span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">" "</span><span class="token punctuation">)</span><span class="token punctuation">,</span> paste<span class="token punctuation">(</span><span class="token string">'/scripts/mofa/mofa2.py'</span><span class="token punctuation">,</span> <span class="token string">"options_MOFA.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span>
</code></pre><h3 id="investigate-mofa-output">Investigate MOFA output</h3><p>We first load the different MOFA models containing the MOFA results
together with the metadata information. Since we don’t know a-priori
which is the ideal number of factors, we can settle on the highest
number of factor where the model stabilises.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Load MOFA output</span>
<span class="token keyword">for</span><span class="token punctuation">(</span>i <span class="token keyword">in</span> c<span class="token punctuation">(</span><span class="token number">5</span><span class="token operator">:</span><span class="token number">15</span><span class="token punctuation">,</span><span class="token number">20</span><span class="token punctuation">,</span>length<span class="token punctuation">(</span>unique<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>sample<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">{</span>
  filepath <span class="token operator">&lt;-</span> paste0<span class="token punctuation">(</span><span class="token string">'results/mofa/mofa_res_'</span><span class="token punctuation">,</span>i<span class="token punctuation">,</span><span class="token string">'factor.hdf5'</span><span class="token punctuation">)</span>
  model <span class="token operator">&lt;-</span> load_model<span class="token punctuation">(</span>filepath<span class="token punctuation">)</span>
  meta_data <span class="token operator">&lt;-</span> read_csv<span class="token punctuation">(</span><span class="token string">"data/metadata/RNA_metadata_cluster.csv"</span><span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
  colnames<span class="token punctuation">(</span>meta_data<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"sample"</span><span class="token punctuation">,</span><span class="token string">"cluster"</span><span class="token punctuation">)</span>
  mofa_metadata <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>samples_metadata<span class="token punctuation">(</span>model<span class="token punctuation">)</span><span class="token punctuation">,</span> meta_data<span class="token punctuation">,</span> by <span class="token operator">=</span> <span class="token string">"sample"</span><span class="token punctuation">)</span>
  names<span class="token punctuation">(</span>mofa_metadata<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"sample"</span><span class="token punctuation">,</span><span class="token string">"group"</span><span class="token punctuation">,</span><span class="token string">"cluster"</span><span class="token punctuation">)</span>
  samples_metadata<span class="token punctuation">(</span>model<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> mofa_metadata
  assign<span class="token punctuation">(</span>paste0<span class="token punctuation">(</span><span class="token string">"model"</span><span class="token punctuation">,</span>i<span class="token punctuation">)</span><span class="token punctuation">,</span>model<span class="token punctuation">)</span>
<span class="token punctuation">}</span>
</code></pre><p>Then, we can compare the factors from different MOFA models and check on
consistency (here: model 7-13).</p><pre class="language-r" tabindex="0"><code class="language-r">compare_factors<span class="token punctuation">(</span>list<span class="token punctuation">(</span>model7<span class="token punctuation">,</span>model8<span class="token punctuation">,</span>model9<span class="token punctuation">,</span>model10<span class="token punctuation">,</span>model11<span class="token punctuation">,</span>model12<span class="token punctuation">,</span>model13<span class="token punctuation">)</span><span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Compare%20MOFA%20models-1.png" alt=""><!-- --></p><p>Here, we can see that the inferred MOFA factor weights do not change
drastically around 9. For the sake of this tutorial, we choose the
9-factor MOFA model.</p><pre class="language-r" tabindex="0"><code class="language-r">model <span class="token operator">&lt;-</span> model10
</code></pre><p>The next part deals with the analysis and interpretation of the MOFA
factors with specific focus on the factor weights. The classic tutorial
on how to perform downstream analysis after MOFA model training can be
found here <a href="https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/downstream_analysis.html">MOFA+: downstream analysis (in
R)</a>.
In this context, it is important to emphasize that the factors can be
interpreted similarly to principal components (PCs) in a principal
component analysis (PCA).</p><p>An initial metric we can explore is the total variance ($R^2$) explained
per view by our MOFA model. This helps us understand how well the model
represents the data.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Investigate factors: Explained total variance per view</span>
plot_variance_explained<span class="token punctuation">(</span>model<span class="token punctuation">,</span> x<span class="token operator">=</span><span class="token string">"group"</span><span class="token punctuation">,</span> y<span class="token operator">=</span><span class="token string">"factor"</span><span class="token punctuation">,</span> plot_total <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">]</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Total%20variance%20per%20factor-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">calculate_variance_explained<span class="token punctuation">(</span>model<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## $r2_total</span>
<span class="token comment">## $single_group</span>
<span class="token comment">##      RNA    metab   proteo </span>
<span class="token comment">## 59.59992 18.70369 23.29474 </span>
<span class="token comment">## </span>
<span class="token comment">## attr(,"class")</span>
<span class="token comment">## [1] "relistable" "list"      </span>
<span class="token comment">## </span>
<span class="token comment">## $r2_per_factor</span>
<span class="token comment">## $single_group</span>
<span class="token comment">##                RNA        metab       proteo</span>
<span class="token comment">## Factor1 24.0201682  0.002979224  0.214741826</span>
<span class="token comment">## Factor2  4.4849965 14.093737072  2.236180823</span>
<span class="token comment">## Factor3  0.2777266  2.193923605 12.875906515</span>
<span class="token comment">## Factor4  7.3683342  2.461682919  4.752507250</span>
<span class="token comment">## Factor5 13.2330359  0.008854678  0.004884149</span>
<span class="token comment">## Factor6  2.3102555  0.001537887  2.692284962</span>
<span class="token comment">## Factor7  4.5839551  0.007795120  0.390045680</span>
<span class="token comment">## Factor8  2.8171650  0.277818450  0.210418704</span>
<span class="token comment">## Factor9  1.9469859  0.004198608  0.003992617</span>
<span class="token comment">## </span>
<span class="token comment">## attr(,"class")</span>
<span class="token comment">## [1] "relistable" "list"</span>
</code></pre><p>The bar plot demonstrates that the nine factors for the view RNA explain
more than 50% of the variance. In contrast, for both metabolomics as
well as proteomics around 20% of the variance can be explained by all
factors.</p><p>It isn’t unexpected that more variance of the RNA dataset can be
reconstructed. Usually, more features available in a given view will
lead to a higher % of variance explained. It is interesting to notice
though that similar amount of variance are explained for metabolic and
proteomic views, despite having very different number of features. This
may be due to the fact that metabolite abundances are in general more
cross-correlated than protein abundances (due to the underlying
regulatory structures, e.i. reaction networks vs protein synthesis and
degratation regulatory mechanisms).</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#Get the cross correlation for RNA and proteomic data</span>
RNA <span class="token operator">&lt;-</span> dcast<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"RNA"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">,</span> feature<span class="token operator">~</span>sample<span class="token punctuation">,</span> value.var <span class="token operator">=</span> <span class="token string">"value"</span><span class="token punctuation">)</span>
row.names<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>
RNA <span class="token operator">&lt;-</span> RNA<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

RNA_crosscor <span class="token operator">&lt;-</span> cor<span class="token punctuation">(</span>t<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">,</span> use <span class="token operator">=</span> <span class="token string">"pairwise.complete.obs"</span><span class="token punctuation">)</span>
<span class="token comment"># hist(RNA_crosscor[upper.tri(RNA_crosscor)], breaks = 1000)</span>

prot <span class="token operator">&lt;-</span> dcast<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"proteo"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">,</span> feature<span class="token operator">~</span>sample<span class="token punctuation">,</span> value.var <span class="token operator">=</span> <span class="token string">"value"</span><span class="token punctuation">)</span>
row.names<span class="token punctuation">(</span>prot<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> prot<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>
prot <span class="token operator">&lt;-</span> prot<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

prot_crosscor <span class="token operator">&lt;-</span> cor<span class="token punctuation">(</span>t<span class="token punctuation">(</span>prot<span class="token punctuation">)</span><span class="token punctuation">,</span> use <span class="token operator">=</span> <span class="token string">"pairwise.complete.obs"</span><span class="token punctuation">)</span>
<span class="token comment"># hist(prot_crosscor[upper.tri(prot_crosscor)], breaks = 1000)</span>

metabo <span class="token operator">&lt;-</span> dcast<span class="token punctuation">(</span>mofa_ready_data<span class="token punctuation">[</span>which<span class="token punctuation">(</span>mofa_ready_data<span class="token operator">$</span>view <span class="token operator">==</span> <span class="token string">"metab"</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">,</span> feature<span class="token operator">~</span>sample<span class="token punctuation">,</span> value.var <span class="token operator">=</span> <span class="token string">"value"</span><span class="token punctuation">)</span>
row.names<span class="token punctuation">(</span>metabo<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> metabo<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>
metabo <span class="token operator">&lt;-</span> metabo<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

metabo_crosscor <span class="token operator">&lt;-</span> cor<span class="token punctuation">(</span>t<span class="token punctuation">(</span>metabo<span class="token punctuation">)</span><span class="token punctuation">,</span> use <span class="token operator">=</span> <span class="token string">"pairwise.complete.obs"</span><span class="token punctuation">)</span>
<span class="token comment"># hist(metabo_crosscor[upper.tri(metabo_crosscor)], breaks = 1000)</span>

<span class="token comment"># plot(calculate_variance_explained(model)$r2_total$single_group, c(mean(RNA_crosscor),mean(metabo_crosscor),mean(prot_crosscor)))</span>
plot<span class="token punctuation">(</span>calculate_variance_explained<span class="token punctuation">(</span>model<span class="token punctuation">)</span><span class="token operator">$</span>r2_total<span class="token operator">$</span>single_group<span class="token punctuation">,</span> c<span class="token punctuation">(</span>dim<span class="token punctuation">(</span>RNA_crosscor<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span>dim<span class="token punctuation">(</span>metabo_crosscor<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span>dim<span class="token punctuation">(</span>prot_crosscor<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Chekc%20cross%20correlation%20of%20omics%20compared%20to%20variance%20epxlained-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">df_variance_crosscor <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>cbind<span class="token punctuation">(</span>calculate_variance_explained<span class="token punctuation">(</span>model<span class="token punctuation">)</span><span class="token operator">$</span>r2_total<span class="token operator">$</span>single_group<span class="token punctuation">,</span>
                                           c<span class="token punctuation">(</span>mean<span class="token punctuation">(</span>RNA_crosscor<span class="token punctuation">)</span><span class="token punctuation">,</span>mean<span class="token punctuation">(</span>metabo_crosscor<span class="token punctuation">)</span><span class="token punctuation">,</span>mean<span class="token punctuation">(</span>prot_crosscor<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
                                           c<span class="token punctuation">(</span>dim<span class="token punctuation">(</span>RNA_crosscor<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span>dim<span class="token punctuation">(</span>metabo_crosscor<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span>dim<span class="token punctuation">(</span>prot_crosscor<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
names<span class="token punctuation">(</span>df_variance_crosscor<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"var_expl"</span><span class="token punctuation">,</span><span class="token string">"mean_crosscor"</span><span class="token punctuation">,</span><span class="token string">"n_features"</span><span class="token punctuation">)</span>
df_variance_crosscor<span class="token operator">$</span>prod <span class="token operator">&lt;-</span> sqrt<span class="token punctuation">(</span>df_variance_crosscor<span class="token operator">$</span>mean_crosscor<span class="token punctuation">)</span> <span class="token operator">*</span> df_variance_crosscor<span class="token operator">$</span>n_features

<span class="token comment">#Just plot them as scatter plot with regression lines</span>
mofa_solved <span class="token operator">&lt;-</span> ggplot<span class="token punctuation">(</span>df_variance_crosscor<span class="token punctuation">,</span> aes<span class="token punctuation">(</span>x <span class="token operator">=</span> prod<span class="token punctuation">,</span> y <span class="token operator">=</span> var_expl<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">+</span> 
  geom_point<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  geom_smooth<span class="token punctuation">(</span>method<span class="token operator">=</span><span class="token string">'lm'</span><span class="token punctuation">,</span> formula<span class="token operator">=</span> y<span class="token operator">~</span>x<span class="token punctuation">)</span> <span class="token operator">+</span>
  theme_minimal<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  xlab<span class="token punctuation">(</span><span class="token string">"sqrt(mean cross-corelation of view) * number of features"</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  ylab<span class="token punctuation">(</span><span class="token string">"variance explained for each view"</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  ggtitle<span class="token punctuation">(</span><span class="token string">"MOFA size and cross cor influence"</span><span class="token punctuation">)</span>

mofa_solved
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Chekc%20cross%20correlation%20of%20omics%20compared%20to%20variance%20epxlained-2.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">mofa_nfeatures <span class="token operator">&lt;-</span> ggplot<span class="token punctuation">(</span>df_variance_crosscor<span class="token punctuation">,</span> aes<span class="token punctuation">(</span>x <span class="token operator">=</span> n_features<span class="token punctuation">,</span> y <span class="token operator">=</span> var_expl<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">+</span> 
  geom_point<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  geom_smooth<span class="token punctuation">(</span>method<span class="token operator">=</span><span class="token string">'lm'</span><span class="token punctuation">,</span> formula<span class="token operator">=</span> y<span class="token operator">~</span>x<span class="token punctuation">)</span> <span class="token operator">+</span>
  theme_minimal<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  xlab<span class="token punctuation">(</span><span class="token string">"sqrt(mean cross-corelation of view) * number of features"</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  ylab<span class="token punctuation">(</span><span class="token string">"variance explained for each view"</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  ggtitle<span class="token punctuation">(</span><span class="token string">"MOFA size influence"</span><span class="token punctuation">)</span>

mofa_nfeatures
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Chekc%20cross%20correlation%20of%20omics%20compared%20to%20variance%20epxlained-3.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#Get the actual coeficient signficance</span>
summary<span class="token punctuation">(</span>lm<span class="token punctuation">(</span>data<span class="token operator">=</span> df_variance_crosscor<span class="token punctuation">,</span> var_expl<span class="token operator">~</span>prod<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## </span>
<span class="token comment">## Call:</span>
<span class="token comment">## lm(formula = var_expl ~ prod, data = df_variance_crosscor)</span>
<span class="token comment">## </span>
<span class="token comment">## Residuals:</span>
<span class="token comment">##      RNA    metab   proteo </span>
<span class="token comment">## -0.06207 -0.77553  0.83760 </span>
<span class="token comment">## </span>
<span class="token comment">## Coefficients:</span>
<span class="token comment">##              Estimate Std. Error t value Pr(&gt;|t|)  </span>
<span class="token comment">## (Intercept) 1.885e+01  8.542e-01   22.06   0.0288 *</span>
<span class="token comment">## prod        1.434e-02  5.176e-04   27.70   0.0230 *</span>
<span class="token comment">## ---</span>
<span class="token comment">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="token comment">## </span>
<span class="token comment">## Residual standard error: 1.143 on 1 degrees of freedom</span>
<span class="token comment">## Multiple R-squared:  0.9987, Adjusted R-squared:  0.9974 </span>
<span class="token comment">## F-statistic: 767.1 on 1 and 1 DF,  p-value: 0.02297</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">summary<span class="token punctuation">(</span>lm<span class="token punctuation">(</span>data<span class="token operator">=</span> df_variance_crosscor<span class="token punctuation">,</span> var_expl<span class="token operator">~</span>n_features<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## </span>
<span class="token comment">## Call:</span>
<span class="token comment">## lm(formula = var_expl ~ n_features, data = df_variance_crosscor)</span>
<span class="token comment">## </span>
<span class="token comment">## Residuals:</span>
<span class="token comment">##    RNA  metab proteo </span>
<span class="token comment">##  1.480  3.426 -4.905 </span>
<span class="token comment">## </span>
<span class="token comment">## Coefficients:</span>
<span class="token comment">##              Estimate Std. Error t value Pr(&gt;|t|)</span>
<span class="token comment">## (Intercept) 14.366419   5.255059   2.734    0.223</span>
<span class="token comment">## n_features   0.007292   0.001446   5.043    0.125</span>
<span class="token comment">## </span>
<span class="token comment">## Residual standard error: 6.163 on 1 degrees of freedom</span>
<span class="token comment">## Multiple R-squared:  0.9622, Adjusted R-squared:  0.9243 </span>
<span class="token comment">## F-statistic: 25.43 on 1 and 1 DF,  p-value: 0.1246</span>
</code></pre><p>Since we would like to use the found correlation for downstream COSMOS
analysis, we first investigate the variance ($R^2$) each factor can
explain per view as well as investigate the factor explaining the most
variance across all views.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Investigate factors: Explained variance per view for each factor </span>
pheatmap<span class="token punctuation">(</span>model<span class="token operator">@</span>cache<span class="token operator">$</span>variance_explained<span class="token operator">$</span>r2_per_factor<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">,</span> display_numbers <span class="token operator">=</span> T<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token string">"0"</span><span class="token punctuation">,</span> legend_labels <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"0"</span><span class="token punctuation">,</span><span class="token string">"10"</span><span class="token punctuation">,</span> <span class="token string">"20"</span><span class="token punctuation">,</span> <span class="token string">"30"</span><span class="token punctuation">,</span> <span class="token string">"40"</span><span class="token punctuation">,</span> <span class="token string">"Variance\n\n"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> legend <span class="token operator">=</span> T<span class="token punctuation">,</span> main <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span> legend_breaks <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token number">0</span><span class="token punctuation">,</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token number">20</span><span class="token punctuation">,</span> <span class="token number">30</span><span class="token punctuation">,</span> <span class="token number">40</span><span class="token punctuation">,</span> max<span class="token punctuation">(</span>model<span class="token operator">@</span>cache<span class="token operator">$</span>variance_explained<span class="token operator">$</span>r2_per_factor<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> color <span class="token operator">=</span> colorRampPalette<span class="token punctuation">(</span>c<span class="token punctuation">(</span><span class="token string">"white"</span><span class="token punctuation">,</span><span class="token string">"red"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">(</span><span class="token number">100</span><span class="token punctuation">)</span><span class="token punctuation">,</span> fontsize_number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Variance%20per%20view%20per%20factor-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">pheatmap<span class="token punctuation">(</span>model<span class="token operator">@</span>cache<span class="token operator">$</span>variance_explained<span class="token operator">$</span>r2_per_factor<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">,</span> display_numbers <span class="token operator">=</span> T<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token string">"0"</span><span class="token punctuation">,</span> legend_labels <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"0"</span><span class="token punctuation">,</span><span class="token string">"10"</span><span class="token punctuation">,</span> <span class="token string">"20"</span><span class="token punctuation">,</span> <span class="token string">"30"</span><span class="token punctuation">,</span> <span class="token string">"40"</span><span class="token punctuation">,</span> <span class="token string">"Variance\n\n"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> legend <span class="token operator">=</span> T<span class="token punctuation">,</span> main <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">,</span> legend_breaks <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token number">0</span><span class="token punctuation">,</span><span class="token number">10</span><span class="token punctuation">,</span> <span class="token number">20</span><span class="token punctuation">,</span> <span class="token number">30</span><span class="token punctuation">,</span> <span class="token number">40</span><span class="token punctuation">,</span> max<span class="token punctuation">(</span>model<span class="token operator">@</span>cache<span class="token operator">$</span>variance_explained<span class="token operator">$</span>r2_per_factor<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> color <span class="token operator">=</span> colorRampPalette<span class="token punctuation">(</span>c<span class="token punctuation">(</span><span class="token string">"white"</span><span class="token punctuation">,</span><span class="token string">"red"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">(</span><span class="token number">100</span><span class="token punctuation">)</span><span class="token punctuation">,</span> fontsize_number <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">,</span>filename <span class="token operator">=</span> <span class="token string">"results/mofa/variance_heatmap.pdf"</span><span class="token punctuation">,</span>width <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">2.5</span><span class="token punctuation">)</span>
</code></pre><p>By analyzing this heatmap, we can observe that Factor 1 and 5 explain a
large proportion of variance of the RNA, Factor 2 and 4 mainly explains
the variance of the metabolomics and Factor 3 highlights variability
from the proteomics view. Consequently, no factor explains a large
portion of the variance across all views, but factor 2 and 4 capture
variance across all views. We can also analyze the correlation between
factors. The next plot highlights the spearman correlation between each
factor.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Investigate factors: Correlation between factors </span>
pdf<span class="token punctuation">(</span>file<span class="token operator">=</span><span class="token string">"results/mofa/plot_factor_cor.pdf"</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">)</span>
plot_factor_cor<span class="token punctuation">(</span>model<span class="token punctuation">,</span> type <span class="token operator">=</span> <span class="token string">'upper'</span><span class="token punctuation">,</span> method <span class="token operator">=</span> <span class="token string">"spearman"</span><span class="token punctuation">,</span> addCoef.col <span class="token operator">=</span> <span class="token string">"black"</span><span class="token punctuation">)</span>
</code></pre><p>Next, we can characterize the factors with the available metadata of
samples. Indeed, it can be interesting to see if some factors are
associated with metadata classes, as they can help to guide our
intepretation of the factors. This part rely on an interpretation of the
Z matrix of the mofa model, while ignoring the factor weight themselves.
The factor weights will be analysed subsequently.</p><p>To begin simply, we can check if some tissue of origin are associated
with any of the factors. There are different ways thsi can be done, but
we find that using a simple linear association between a given tissues
category in a factor compared to any other tissue categories
(e.g.&nbsp;breast tissue vs every other tissue types) provide an easy to
interpret insight that serves our purpose well. To do so, we use the
Univariate Linear Model (ULM) function of decoupleR.</p><pre class="language-r" tabindex="0"><code class="language-r">NCI_60_metadata <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>read_csv<span class="token punctuation">(</span>file <span class="token operator">=</span> <span class="token string">"data/metadata/RNA_metadata_cluster.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## Rows: 60 Columns: 16</span>
<span class="token comment">## ── Column specification ────────────────────────────────────────────────────────</span>
<span class="token comment">## Delimiter: ","</span>
<span class="token comment">## chr (12): cell_line, tissue of origin a, sex a, prior treatment a,b, Epithel...</span>
<span class="token comment">## dbl  (4): cluster, age a, mdr f, doubling time g</span>
<span class="token comment">## </span>
<span class="token comment">## ℹ Use `spec()` to retrieve the full column specification for this data.</span>
<span class="token comment">## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#discretize age category so that regression can be applied</span>
NCI_60_metadata<span class="token operator">$</span>`age over <span class="token number">50</span>` <span class="token operator">&lt;-</span> NCI_60_metadata<span class="token operator">$</span>`age a` <span class="token operator">&gt;</span> <span class="token number">50</span>
tissues <span class="token operator">&lt;-</span>NCI_60_metadata<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
names<span class="token punctuation">(</span>tissues<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"source"</span><span class="token punctuation">,</span><span class="token string">"target"</span><span class="token punctuation">)</span>

Z_matrix <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>model<span class="token operator">@</span>expectations<span class="token operator">$</span>Z<span class="token operator">$</span>single_group<span class="token punctuation">)</span>

<span class="token comment">#Check specifically tissue meta-data association with Z matrix</span>
tissue_enrichment <span class="token operator">&lt;-</span> decoupleR<span class="token operator">::</span>run_ulm<span class="token punctuation">(</span>Z_matrix<span class="token punctuation">,</span> tissues<span class="token punctuation">)</span>
tissue_enrichment <span class="token operator">&lt;-</span> reshape2<span class="token operator">::</span>dcast<span class="token punctuation">(</span>tissue_enrichment<span class="token punctuation">,</span> source<span class="token operator">~</span>condition<span class="token punctuation">,</span> value.var <span class="token operator">=</span> <span class="token string">"score"</span><span class="token punctuation">)</span>
row.names<span class="token punctuation">(</span>tissue_enrichment<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> tissue_enrichment<span class="token operator">$</span>source
tissue_enrichment <span class="token operator">&lt;-</span> tissue_enrichment<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>

<span class="token comment">#plot them as heatmaps</span>
palette <span class="token operator">&lt;-</span> make_heatmap_color_palette<span class="token punctuation">(</span>tissue_enrichment<span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>t<span class="token punctuation">(</span>tissue_enrichment<span class="token punctuation">)</span><span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span>display_numbers <span class="token operator">=</span> F<span class="token punctuation">,</span> filename <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_tissue_enrichment.pdf"</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">2.6</span><span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>tissue_enrichment<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span>display_numbers <span class="token operator">=</span> T<span class="token punctuation">)</span>
</code></pre><p>Here, we can see that factor 4 seems to be strongly associated with
eukemia, and factor 2 with melanoma.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#format meta-data names</span>
all_metadata <span class="token operator">&lt;-</span> reshape2<span class="token operator">::</span>melt<span class="token punctuation">(</span>NCI_60_metadata<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">17</span><span class="token punctuation">,</span><span class="token number">5</span><span class="token punctuation">,</span><span class="token number">7</span><span class="token punctuation">,</span><span class="token number">8</span><span class="token punctuation">,</span><span class="token number">10</span><span class="token punctuation">,</span><span class="token number">13</span><span class="token punctuation">,</span><span class="token number">14</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span> id.vars <span class="token operator">=</span> <span class="token string">"cell_line"</span><span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>variable <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">" a$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>variable<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>variable <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">" a,c$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>variable<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>variable <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">" d$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>variable<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>variable <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"histology"</span><span class="token punctuation">,</span><span class="token string">"histo"</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>variable<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>variable <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"tissue of origin"</span><span class="token punctuation">,</span><span class="token string">"tissue"</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>variable<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>value <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"[(]81-103[)]"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>value<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>value <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"[(]35-57[)]"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>value<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>value <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"Memorial Sloan Kettering Cancer Center"</span><span class="token punctuation">,</span><span class="token string">"Memorial SKCC"</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>value<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>value <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"Malignant melanotic melanoma"</span><span class="token punctuation">,</span><span class="token string">"Malig. melanotic melan."</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>value<span class="token punctuation">)</span>
all_metadata<span class="token operator">$</span>value <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"Ductal carcinoma-mammary gland"</span><span class="token punctuation">,</span><span class="token string">"Duct. carci-mam. gland"</span><span class="token punctuation">,</span>all_metadata<span class="token operator">$</span>value<span class="token punctuation">)</span>

all_metadata<span class="token operator">$</span>target <span class="token operator">&lt;-</span> paste<span class="token punctuation">(</span>all_metadata<span class="token operator">$</span>variable<span class="token punctuation">,</span> all_metadata<span class="token operator">$</span>value<span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">": "</span><span class="token punctuation">)</span>
all_metadata <span class="token operator">&lt;-</span> all_metadata<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
names<span class="token punctuation">(</span>all_metadata<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"source"</span><span class="token punctuation">,</span><span class="token string">"target"</span><span class="token punctuation">)</span>
  
Z_matrix <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>model<span class="token operator">@</span>expectations<span class="token operator">$</span>Z<span class="token operator">$</span>single_group<span class="token punctuation">)</span>

<span class="token comment">#run ULM on the Z matrix meta data to find associated metadata features</span>
all_metadata_enrichment <span class="token operator">&lt;-</span> decoupleR<span class="token operator">::</span>run_ulm<span class="token punctuation">(</span>Z_matrix<span class="token punctuation">,</span> all_metadata<span class="token punctuation">,</span> minsize <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span>
all_metadata_enrichment <span class="token operator">&lt;-</span> reshape2<span class="token operator">::</span>dcast<span class="token punctuation">(</span>all_metadata_enrichment<span class="token punctuation">,</span> source<span class="token operator">~</span>condition<span class="token punctuation">,</span> value.var <span class="token operator">=</span> <span class="token string">"score"</span><span class="token punctuation">)</span>
row.names<span class="token punctuation">(</span>all_metadata_enrichment<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> all_metadata_enrichment<span class="token operator">$</span>source
all_metadata_enrichment <span class="token operator">&lt;-</span> all_metadata_enrichment<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">]</span>
all_metadata_enrichment_top <span class="token operator">&lt;-</span> all_metadata_enrichment<span class="token punctuation">[</span>
  apply<span class="token punctuation">(</span>all_metadata_enrichment<span class="token punctuation">,</span> <span class="token number">1</span><span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">{</span>max<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">&gt;</span> <span class="token number">2</span><span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
<span class="token punctuation">]</span>

<span class="token comment">#Plot the heatmap</span>
palette <span class="token operator">&lt;-</span> make_heatmap_color_palette<span class="token punctuation">(</span>all_metadata_enrichment_top<span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>t<span class="token punctuation">(</span>all_metadata_enrichment_top<span class="token punctuation">)</span><span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span>display_numbers <span class="token operator">=</span> F<span class="token punctuation">,</span> filename <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_metadata_enrichment.pdf"</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">4.5</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>t<span class="token punctuation">(</span>all_metadata_enrichment_top<span class="token punctuation">)</span><span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span>display_numbers <span class="token operator">=</span> T<span class="token punctuation">)</span>
</code></pre><p>Further, we can inspect the composition of the factors by plotting the
top 10 weights per factor and view.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">## Top 20 factor weights per view</span>
grid.arrange<span class="token punctuation">(</span>
  <span class="token comment">### RNA</span>
  plot_top_weights<span class="token punctuation">(</span>model<span class="token punctuation">,</span> 
                   view <span class="token operator">=</span> <span class="token string">"RNA"</span><span class="token punctuation">,</span>
                   factor <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span>
                   nfeatures <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span> <span class="token operator">+</span>
    ggtitle<span class="token punctuation">(</span><span class="token string">"RNA"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token comment">### Proteomics</span>
  plot_top_weights<span class="token punctuation">(</span>model<span class="token punctuation">,</span> 
                   view <span class="token operator">=</span> <span class="token string">"proteo"</span><span class="token punctuation">,</span>
                   factor <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span>
                   nfeatures <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span> <span class="token operator">+</span> 
    ggtitle<span class="token punctuation">(</span><span class="token string">"Proteomics"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token comment">### Metabolomics</span>
  plot_top_weights<span class="token punctuation">(</span>model<span class="token punctuation">,</span> 
                   view <span class="token operator">=</span> <span class="token string">"metab"</span><span class="token punctuation">,</span>
                   factor <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span>
                   nfeatures <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span> <span class="token operator">+</span> 
    ggtitle<span class="token punctuation">(</span><span class="token string">"Metabolomics"</span><span class="token punctuation">)</span><span class="token punctuation">,</span> ncol <span class="token operator">=</span><span class="token number">3</span>
<span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Factor%20weights%20per%20view-1.png" alt=""><!-- --></p><p>Using this visualization, features with strong association with the
factor (large absolute values) can be easily identified and further
inspected by literature investigation.</p><p>Further tools to investigate the latent factors are available inside the
MOFA framework (<a href="https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/downstream_analysis.html">MOFA+: downstream analysis (in
R)</a>).</p><h3 id="from-mofa-output-to-tf-activity-ligand-receptor-activity">From MOFA output to TF activity, ligand-receptor activity</h3><p>In the next step, using the weights of Factor 4 (highest shared $R^2$
across views), different databases (liana and dorothea) and decoupleR,
the transcription factor activities, ligand-receptor scores are
estimated.</p><p>First, we have to extract the weights from our MOFA model and keep the
weights from Factor 4 in the RNA view.</p><pre class="language-r" tabindex="0"><code class="language-r">weights <span class="token operator">&lt;-</span> get_weights<span class="token punctuation">(</span>model<span class="token punctuation">,</span> views <span class="token operator">=</span> <span class="token string">"all"</span><span class="token punctuation">,</span> factors <span class="token operator">=</span> <span class="token string">"all"</span><span class="token punctuation">)</span>

save<span class="token punctuation">(</span>weights<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_weights.RData"</span><span class="token punctuation">)</span>
<span class="token comment"># Extract Factor with highest explained variance in RNA view</span>
RNA <span class="token operator">&lt;-</span> data.frame<span class="token punctuation">(</span>weights<span class="token operator">$</span>RNA<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">]</span><span class="token punctuation">)</span> 
row.names<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_RNA"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><p>Then we load the consensus networks from our databases, as well as the
TF-targets regulons from collectri.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Load LIANA (receptor and ligand) consensus network</span>
<span class="token comment"># ligrec_ressource &lt;- distinct(liana::decomplexify(liana::select_resource("Consensus")[[1]]))</span>
<span class="token comment"># save(ligrec_ressource, file = "support/ligrec_ressource.RData")</span>

<span class="token comment">#process the LR interaction prior knowledge to make suitable input for decoupleR</span>
load<span class="token punctuation">(</span>file <span class="token operator">=</span> <span class="token string">"support/ligrec_ressource.RData"</span><span class="token punctuation">)</span>
ligrec_geneset <span class="token operator">&lt;-</span> format_LR_ressource<span class="token punctuation">(</span>ligrec_ressource<span class="token punctuation">)</span>

<span class="token comment"># Load Dorothea (TF) network</span>
<span class="token comment"># dorothea_df &lt;- decoupleR::get_collectri()</span>
<span class="token comment"># save(dorothea_df, file = "support/dorothea_df.RData")</span>
load<span class="token punctuation">(</span>file <span class="token operator">=</span> <span class="token string">"support/dorothea_df.RData"</span><span class="token punctuation">)</span>
</code></pre><p>We can display what are the top weights of the model and to which factor
they are associated.</p><pre class="language-r" tabindex="0"><code class="language-r">RNA_all <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>weights<span class="token operator">$</span>RNA<span class="token punctuation">)</span> 
row.names<span class="token punctuation">(</span>RNA_all<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_RNA"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>RNA_all<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment">#Extract the top weights</span>
RNA_top <span class="token operator">&lt;-</span> RNA_all<span class="token punctuation">[</span>order<span class="token punctuation">(</span>apply<span class="token punctuation">(</span>RNA_all<span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">{</span>max<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">25</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token punctuation">]</span>

<span class="token comment">#plot them as heatmaps</span>
palette <span class="token operator">&lt;-</span> make_heatmap_color_palette<span class="token punctuation">(</span>RNA_top<span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>RNA_top<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span> filename <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_top_RNA.pdf"</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">4.3</span><span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>RNA_top<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">)</span>
</code></pre><p>We can do the same for metabolites</p><pre class="language-r" tabindex="0"><code class="language-r">metabs_all <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>weights<span class="token operator">$</span>metab<span class="token punctuation">)</span> 
row.names<span class="token punctuation">(</span>metabs_all<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_metab"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>metabs_all<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment">#extract top metab weights</span>
metabs_top <span class="token operator">&lt;-</span> metabs_all<span class="token punctuation">[</span>order<span class="token punctuation">(</span>apply<span class="token punctuation">(</span>metabs_all<span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">{</span>max<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token punctuation">]</span>

<span class="token comment">#plot them as heatmaps</span>
palette <span class="token operator">&lt;-</span> make_heatmap_color_palette<span class="token punctuation">(</span>metabs_top<span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>metabs_top<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span> filename <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_top_metabs.pdf"</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">2.2</span><span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>metabs_top<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">)</span>
</code></pre><p>Then, by using decoupleR and prior knowledge networks, we calculate the
different regulatory activities inferred by the ULM approach. Depending
on the task, the minimum number of targets per source varies required to
maintain the source in the output (e.g.&nbsp;two targets by definition for
ligand-receptor interactions). Here can display them as heatmaps.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Calculate regulatory activities from Receptor and Ligand network</span>
ligrec_factors <span class="token operator">&lt;-</span> run_ulm<span class="token punctuation">(</span>mat <span class="token operator">=</span> as.matrix<span class="token punctuation">(</span>RNA_all<span class="token punctuation">)</span><span class="token punctuation">,</span> network <span class="token operator">=</span> ligrec_geneset<span class="token punctuation">,</span> .source <span class="token operator">=</span> set<span class="token punctuation">,</span> .target <span class="token operator">=</span> gene<span class="token punctuation">,</span> minsize <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">)</span> 
ligrec_factors_df <span class="token operator">&lt;-</span> wide_ulm_res<span class="token punctuation">(</span>ligrec_factors<span class="token punctuation">)</span>
ligrec_factors_df <span class="token operator">&lt;-</span> ligrec_factors_df<span class="token punctuation">[</span>order<span class="token punctuation">(</span>apply<span class="token punctuation">(</span>ligrec_factors_df<span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">{</span>max<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">25</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token punctuation">]</span>


<span class="token comment">#plot them as heatmaps</span>
palette <span class="token operator">&lt;-</span> make_heatmap_color_palette<span class="token punctuation">(</span>ligrec_factors_df<span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>ligrec_factors_df<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span> filename <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_top_ligrec.pdf"</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">4.3</span><span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>ligrec_factors_df<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">)</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Calculate regulatory activities from TF network</span>
TF_factors <span class="token operator">&lt;-</span> run_ulm<span class="token punctuation">(</span>mat <span class="token operator">=</span> as.matrix<span class="token punctuation">(</span>RNA_all<span class="token punctuation">)</span><span class="token punctuation">,</span> network <span class="token operator">=</span> dorothea_df<span class="token punctuation">,</span> minsize <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span>
TF_factors <span class="token operator">&lt;-</span> wide_ulm_res<span class="token punctuation">(</span>TF_factors<span class="token punctuation">)</span>
TF_factors <span class="token operator">&lt;-</span> TF_factors<span class="token punctuation">[</span>order<span class="token punctuation">(</span>apply<span class="token punctuation">(</span>TF_factors<span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">{</span>max<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>x<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">25</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token punctuation">]</span>

<span class="token comment">#Plot them as heatmaps</span>
palette <span class="token operator">&lt;-</span> make_heatmap_color_palette<span class="token punctuation">(</span>TF_factors<span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>TF_factors<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span> filename <span class="token operator">=</span> <span class="token string">"results/mofa/mofa_top_TF.pdf"</span><span class="token punctuation">,</span> width <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span> height <span class="token operator">=</span> <span class="token number">4.3</span><span class="token punctuation">)</span>
pheatmap<span class="token punctuation">(</span>TF_factors<span class="token punctuation">,</span> show_rownames <span class="token operator">=</span> T<span class="token punctuation">,</span> cluster_cols <span class="token operator">=</span> F<span class="token punctuation">,</span> cluster_rows <span class="token operator">=</span> F<span class="token punctuation">,</span>color <span class="token operator">=</span> palette<span class="token punctuation">,</span> angle_col <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">)</span>
</code></pre><p>These activity estimations are saved in an input file for further
downstream analysis as well.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Calculate regulatory activities from Receptor and Ligand network</span>
ligrec_high_vs_low <span class="token operator">&lt;-</span> run_ulm<span class="token punctuation">(</span>mat <span class="token operator">=</span> as.matrix<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">,</span> network <span class="token operator">=</span> ligrec_geneset<span class="token punctuation">,</span> .source <span class="token operator">=</span> set<span class="token punctuation">,</span> .target <span class="token operator">=</span> gene<span class="token punctuation">,</span> minsize <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">)</span> 
ligrec_high_vs_low <span class="token operator">&lt;-</span> ligrec_high_vs_low<span class="token punctuation">[</span>ligrec_high_vs_low<span class="token operator">$</span>statistic <span class="token operator">==</span> <span class="token string">"ulm"</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
ligrec_high_vs_low_vector <span class="token operator">&lt;-</span> ligrec_high_vs_low<span class="token operator">$</span>score
names<span class="token punctuation">(</span>ligrec_high_vs_low_vector<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> ligrec_high_vs_low<span class="token operator">$</span>source

ligrec_high_vs_low_top <span class="token operator">&lt;-</span> ligrec_high_vs_low<span class="token punctuation">[</span>order<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>ligrec_high_vs_low<span class="token operator">$</span>score<span class="token punctuation">)</span><span class="token punctuation">,</span>decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">15</span><span class="token punctuation">]</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
ligrec_high_vs_low_top<span class="token operator">$</span>source <span class="token operator">&lt;-</span> factor<span class="token punctuation">(</span>ligrec_high_vs_low_top<span class="token operator">$</span>source<span class="token punctuation">,</span> levels <span class="token operator">=</span> ligrec_high_vs_low_top<span class="token operator">$</span>source<span class="token punctuation">)</span>
ggplot<span class="token punctuation">(</span>ligrec_high_vs_low_top<span class="token punctuation">,</span> aes<span class="token punctuation">(</span>x<span class="token operator">=</span>source<span class="token punctuation">,</span> y <span class="token operator">=</span> score<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">+</span> geom_bar<span class="token punctuation">(</span>stat<span class="token operator">=</span> <span class="token string">"identity"</span><span class="token punctuation">,</span>position <span class="token operator">=</span> <span class="token string">"dodge"</span><span class="token punctuation">)</span> <span class="token operator">+</span> theme_minimal<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span> theme<span class="token punctuation">(</span>axis.text.x <span class="token operator">=</span> element_text<span class="token punctuation">(</span>angle <span class="token operator">=</span> <span class="token number">315</span><span class="token punctuation">,</span> vjust <span class="token operator">=</span> <span class="token number">0.5</span><span class="token punctuation">,</span> hjust<span class="token operator">=</span><span class="token number">0</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Activity%20estimations%20for%20factor%204-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Calculate regulatory activities from TF network</span>
TF_high_vs_low <span class="token operator">&lt;-</span> run_ulm<span class="token punctuation">(</span>mat <span class="token operator">=</span> as.matrix<span class="token punctuation">(</span>RNA<span class="token punctuation">)</span><span class="token punctuation">,</span> network <span class="token operator">=</span> dorothea_df<span class="token punctuation">,</span> minsize <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">)</span>
TF_high_vs_low <span class="token operator">&lt;-</span> TF_high_vs_low<span class="token punctuation">[</span>TF_high_vs_low<span class="token operator">$</span>statistic <span class="token operator">==</span> <span class="token string">"ulm"</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
TF_high_vs_low_vector <span class="token operator">&lt;-</span> TF_high_vs_low<span class="token operator">$</span>score
names<span class="token punctuation">(</span>TF_high_vs_low_vector<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> TF_high_vs_low<span class="token operator">$</span>source

<span class="token comment"># Prepare moon analysis input</span>
TF_high_vs_low <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>TF_high_vs_low<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
row.names<span class="token punctuation">(</span>TF_high_vs_low<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> TF_high_vs_low<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span>

TF_high_vs_low_top_10 <span class="token operator">&lt;-</span> TF_high_vs_low<span class="token punctuation">[</span>order<span class="token punctuation">(</span>abs<span class="token punctuation">(</span>TF_high_vs_low<span class="token operator">$</span>score<span class="token punctuation">)</span><span class="token punctuation">,</span> decreasing <span class="token operator">=</span> T<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
TF_high_vs_low_top_10<span class="token operator">$</span>source <span class="token operator">&lt;-</span> factor<span class="token punctuation">(</span>TF_high_vs_low_top_10<span class="token operator">$</span>source<span class="token punctuation">,</span> levels <span class="token operator">=</span> TF_high_vs_low_top_10<span class="token operator">$</span>source<span class="token punctuation">)</span>
ggplot<span class="token punctuation">(</span>TF_high_vs_low_top_10<span class="token punctuation">,</span> aes<span class="token punctuation">(</span>x<span class="token operator">=</span>source<span class="token punctuation">,</span> y <span class="token operator">=</span> score<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">+</span> geom_bar<span class="token punctuation">(</span>stat<span class="token operator">=</span> <span class="token string">"identity"</span><span class="token punctuation">,</span>position <span class="token operator">=</span> <span class="token string">"dodge"</span><span class="token punctuation">)</span> <span class="token operator">+</span> theme_minimal<span class="token punctuation">(</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Activity%20estimations%20for%20factor%204-2.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Combine results</span>
ligrec_TF_moon_inputs <span class="token operator">&lt;-</span> list<span class="token punctuation">(</span><span class="token string">"ligrec"</span> <span class="token operator">=</span> ligrec_high_vs_low_vector<span class="token punctuation">,</span>
                              <span class="token string">"TF"</span> <span class="token operator">=</span> TF_high_vs_low_vector<span class="token punctuation">)</span>

save<span class="token punctuation">(</span>ligrec_TF_moon_inputs<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"data/cosmos/ligrec_TF_moon_inputs.Rdata"</span><span class="token punctuation">)</span>
</code></pre><p>We have now successfully used the MOFA weights (Factor 4, RNA view) to
infer not only TF activities but also the ligand-receptor interactions.</p><h3 id="extracting-network-with-cosmos-to-formulate-mechanistic-hypotheses">Extracting network with COSMOS to formulate mechanistic hypotheses</h3><p>In this next parts the COSMOS run is going to be performed. We can use
the output of decoupleR (the activities) next to the factor weights as
an input for COSMOS to specifically focus the analysis on pre-computed
data-driven results.</p><p>Here, in the first step, the data is loaded. Besides the activity
estimations and the factor weights, the previous filtered out expressed
gene names are loaded.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># Load data</span>
load<span class="token punctuation">(</span>file <span class="token operator">=</span> <span class="token string">"data/cosmos/ligrec_TF_moon_inputs.Rdata"</span><span class="token punctuation">)</span>

signaling_input <span class="token operator">&lt;-</span> ligrec_TF_moon_inputs<span class="token operator">$</span>TF
ligrec_input <span class="token operator">&lt;-</span> ligrec_TF_moon_inputs<span class="token operator">$</span>ligrec

RNA_input <span class="token operator">&lt;-</span> weights<span class="token operator">$</span>RNA<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">]</span> 
prot_input <span class="token operator">&lt;-</span> weights<span class="token operator">$</span>proteo<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">]</span>
metab_inputs <span class="token operator">&lt;-</span> weights<span class="token operator">$</span>metab<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">]</span>

<span class="token comment">#remove MOFA tags from feature names</span>
names<span class="token punctuation">(</span>RNA_input<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_RNA"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>RNA_input<span class="token punctuation">)</span><span class="token punctuation">)</span>
names<span class="token punctuation">(</span>prot_input<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_proteo"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>prot_input<span class="token punctuation">)</span><span class="token punctuation">)</span>

RNA_log2_FPKM <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>read_csv<span class="token punctuation">(</span><span class="token string">"data/RNA/RNA_log2_FPKM_clean.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token operator">$</span>Genes <span class="token comment">#since only complete cases were considered in the first part, here the full gene list is loaded</span>
</code></pre><p>We can be interested in looking how RNA and corresponding proteins
correlate in each factor. We can plot this using scatter plots for each
factors.</p><pre class="language-r" tabindex="0"><code class="language-r">weights_RNA <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>weights<span class="token operator">$</span>RNA<span class="token punctuation">)</span>
weights_prot <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>weights<span class="token operator">$</span>proteo<span class="token punctuation">)</span>

weights_prot<span class="token operator">$</span>ID <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_proteo$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>weights_prot<span class="token punctuation">)</span><span class="token punctuation">)</span>
weights_RNA<span class="token operator">$</span>ID <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_RNA$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>row.names<span class="token punctuation">(</span>weights_RNA<span class="token punctuation">)</span><span class="token punctuation">)</span>

plot_list <span class="token operator">&lt;-</span> list<span class="token punctuation">(</span><span class="token punctuation">)</span>
r2_list <span class="token operator">&lt;-</span> list<span class="token punctuation">(</span><span class="token punctuation">)</span>
<span class="token keyword">for</span><span class="token punctuation">(</span>i <span class="token keyword">in</span> <span class="token number">1</span><span class="token operator">:</span><span class="token punctuation">(</span>dim<span class="token punctuation">(</span>weights_prot<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">{</span>
  merged_weights <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>weights_RNA<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span>i<span class="token punctuation">,</span><span class="token number">10</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span>weights_prot<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span>i<span class="token punctuation">,</span><span class="token number">10</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">,</span> by <span class="token operator">=</span> <span class="token string">"ID"</span><span class="token punctuation">)</span>
  names<span class="token punctuation">(</span>merged_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"ID"</span><span class="token punctuation">,</span><span class="token string">"RNA"</span><span class="token punctuation">,</span><span class="token string">"prot"</span><span class="token punctuation">)</span>
  r2_list<span class="token punctuation">[</span><span class="token punctuation">[</span>i<span class="token punctuation">]</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> cor<span class="token punctuation">(</span>merged_weights<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">,</span>merged_weights<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token operator">^</span><span class="token number">2</span>
  plot_list<span class="token punctuation">[</span><span class="token punctuation">[</span>i<span class="token punctuation">]</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> ggplot<span class="token punctuation">(</span>merged_weights<span class="token punctuation">,</span>aes<span class="token punctuation">(</span>x <span class="token operator">=</span> RNA<span class="token punctuation">,</span>y <span class="token operator">=</span> prot<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token operator">+</span> geom_point<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span>
  geom_smooth<span class="token punctuation">(</span>method<span class="token operator">=</span><span class="token string">'lm'</span><span class="token punctuation">,</span> formula<span class="token operator">=</span> y<span class="token operator">~</span>x<span class="token punctuation">)</span> <span class="token operator">+</span> 
  theme_minimal<span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span> 
  theme<span class="token punctuation">(</span>axis.title.x<span class="token operator">=</span>element_blank<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
      axis.text.x<span class="token operator">=</span>element_blank<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
      axis.ticks.x<span class="token operator">=</span>element_blank<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
      axis.title.y<span class="token operator">=</span>element_blank<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
      axis.text.y<span class="token operator">=</span>element_blank<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
      axis.ticks.y<span class="token operator">=</span>element_blank<span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span>

r2_vec <span class="token operator">&lt;-</span> unlist<span class="token punctuation">(</span>r2_list<span class="token punctuation">)</span>
r2_vec
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] 0.041804090 0.411093174 0.038741311 0.421953345 0.002666596 0.291243355</span>
<span class="token comment">## [7] 0.027585257 0.058177119 0.019215245</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">ggsave<span class="token punctuation">(</span>filename <span class="token operator">=</span> <span class="token string">"results/mofa/RNA_prot_cor.pdf"</span><span class="token punctuation">,</span>plot <span class="token operator">=</span> do.call<span class="token punctuation">(</span><span class="token string">"grid.arrange"</span><span class="token punctuation">,</span> c<span class="token punctuation">(</span>plot_list<span class="token punctuation">,</span> ncol <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> device <span class="token operator">=</span> <span class="token string">"pdf"</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## Saving 7 x 7 in image</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/correlation%20RNA/prot-1.png" alt=""><!-- -->
Coherently, only the factors where there is variance explained for both
RNA and proteins are correlated.</p><p>Next, we perform filtering steps in order to identify top deregulated
features. Here, the individual filtering is based on absolute factor
weight or activity score. The first step involves the visualization of
the respective distribution to define an appropriate threshold.</p><p>First, the RNA factor weights are filtered based on a threshold defined
by their distribution as well a threshold defined by the distribution of
the protein factor weights.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">##RNA</span>
<span class="token punctuation">{</span>plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>RNA_input<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">0.2</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">0.2</span><span class="token punctuation">)</span><span class="token punctuation">}</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Plot:%20RNA%20weights%20&amp;%20protein%20weights-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token punctuation">{</span>plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>prot_input<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">0.05</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">0.05</span><span class="token punctuation">)</span><span class="token punctuation">}</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Plot:%20RNA%20weights%20&amp;%20protein%20weights-2.png" alt=""><!-- --></p><p>Based on the plots and indicated by the straight lines, the RNA weights
threshold is set to -0.2 and 0.2, and the protein weights threshold to
-0.05 and 0.05. RNA factor weight values lying inside this threshold are
set to 0 and previously filtered out genes (before MOFA analysis) are
also included with value 0.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token keyword">for</span><span class="token punctuation">(</span>gene <span class="token keyword">in</span> names<span class="token punctuation">(</span>RNA_input<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
  <span class="token keyword">if</span> <span class="token punctuation">(</span>RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&gt;</span> <span class="token operator">-</span><span class="token number">0.2</span> <span class="token operator">&amp;</span> RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&lt;</span> <span class="token number">0.2</span><span class="token punctuation">)</span>
  <span class="token punctuation">{</span>
    RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token number">0</span>
  <span class="token punctuation">}</span> <span class="token keyword">else</span>
  <span class="token punctuation">{</span>
    RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&lt;-</span> sign<span class="token punctuation">(</span>RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span><span class="token punctuation">)</span> <span class="token operator">*</span> <span class="token number">10</span>
  <span class="token punctuation">}</span>
  <span class="token keyword">if</span> <span class="token punctuation">(</span>gene <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>prot_input<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token comment">#will need to add a sign consistency check between RNA and prot as well</span>
  <span class="token punctuation">{</span>
    <span class="token keyword">if</span> <span class="token punctuation">(</span>prot_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&gt;</span> <span class="token operator">-</span><span class="token number">0.05</span> <span class="token operator">&amp;</span> prot_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&lt;</span> <span class="token number">0.05</span><span class="token punctuation">)</span>
    <span class="token punctuation">{</span>
      RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token number">0</span>
    <span class="token punctuation">}</span> <span class="token keyword">else</span>
    <span class="token punctuation">{</span>
      RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span> <span class="token operator">&lt;-</span> sign<span class="token punctuation">(</span>RNA_input<span class="token punctuation">[</span>gene<span class="token punctuation">]</span><span class="token punctuation">)</span> <span class="token operator">*</span> <span class="token number">10</span>
    <span class="token punctuation">}</span>
  <span class="token punctuation">}</span>
<span class="token punctuation">}</span>

RNA_log2_FPKM <span class="token operator">&lt;-</span> RNA_log2_FPKM<span class="token punctuation">[</span><span class="token operator">!</span><span class="token punctuation">(</span>RNA_log2_FPKM <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>RNA_input<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
genes <span class="token operator">&lt;-</span> RNA_log2_FPKM
RNA_log2_FPKM <span class="token operator">&lt;-</span> rep<span class="token punctuation">(</span><span class="token number">0</span><span class="token punctuation">,</span>length<span class="token punctuation">(</span>RNA_log2_FPKM<span class="token punctuation">)</span><span class="token punctuation">)</span>
names<span class="token punctuation">(</span>RNA_log2_FPKM<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> genes

RNA_input <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span>RNA_input<span class="token punctuation">,</span> RNA_log2_FPKM<span class="token punctuation">)</span>
</code></pre><p>The same procedure is repeated for the transcription factor activities.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">## Transcription factors</span>
<span class="token punctuation">{</span>plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>signaling_input<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">0.5</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">3.5</span><span class="token punctuation">)</span><span class="token punctuation">}</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Plot:%20TF%20activity%20estimates-1.png" alt=""><!-- --></p><p>The threshold is set to -0.5 and 3.5 respectively. Further, the
TF_to_remove variable is later used to remove transcription factors with
activities below this threshold from the prior knowledge network.</p><pre class="language-r" tabindex="0"><code class="language-r">TF_to_remove <span class="token operator">&lt;-</span> signaling_input<span class="token punctuation">[</span>signaling_input <span class="token operator">&gt;</span> <span class="token operator">-</span><span class="token number">0.5</span> <span class="token operator">&amp;</span> signaling_input <span class="token operator">&lt;</span> <span class="token number">3.5</span><span class="token punctuation">]</span>
signaling_input <span class="token operator">&lt;-</span> signaling_input<span class="token punctuation">[</span>signaling_input <span class="token operator">&lt;</span> <span class="token operator">-</span><span class="token number">0.5</span> <span class="token operator">|</span> signaling_input <span class="token operator">&gt;</span> <span class="token number">3.5</span><span class="token punctuation">]</span>
</code></pre><p>The same procedure is repeated for the ligand-receptor activities.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">## Ligand-receptor interactions</span>
<span class="token punctuation">{</span>plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>ligrec_input<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">0.5</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">2.5</span><span class="token punctuation">)</span><span class="token punctuation">}</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Plot:%20Ligand-receptor%20activity%20estimates-1.png" alt=""><!-- --></p><p>Here, only activities higher than 2.5 or lower than -0.5 are kept (see
straight lines in plot). This is an arbitrary theshold aimed at keeping
a relativelly high number of interesting fetures to connect in the rest
of the pipeline. Since at this point we are interested in the receptors,
the names are adjusted accordingly and if there are multiple mentions of
a receptor, its activity is calculated by the mean of the associated
ligand-receptor interactions.</p><pre class="language-r" tabindex="0"><code class="language-r">ligrec_input <span class="token operator">&lt;-</span> ligrec_input<span class="token punctuation">[</span>ligrec_input <span class="token operator">&gt;</span> <span class="token number">2.5</span> <span class="token operator">|</span> ligrec_input <span class="token operator">&lt;</span> <span class="token operator">-</span><span class="token number">0.5</span><span class="token punctuation">]</span>

rec_inputs <span class="token operator">&lt;-</span> ligrec_input
names<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">".+___"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span><span class="token punctuation">)</span>
rec_inputs <span class="token operator">&lt;-</span> tapply<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">,</span> names<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span><span class="token punctuation">,</span> mean<span class="token punctuation">)</span>
receptors <span class="token operator">&lt;-</span> names<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span>
rec_inputs <span class="token operator">&lt;-</span> as.numeric<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> receptors
</code></pre><p>Finally, the same procedure is repeated for the metabolite factor
weights.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">## Metabolites</span>
<span class="token punctuation">{</span>plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">0.2</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">0.2</span><span class="token punctuation">)</span><span class="token punctuation">}</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Plot:%20Metabolite%20factor%20weights-1.png" alt=""><!-- --></p><p>The threshold is set to 0.2 and -0.2 respectively. Further, the
metab_to_exclude variable is later used to remove metabolites with
activities below this threshold from the prior knowledge network.
Moreover, metabolite names are translated into HMDB format and metabolic
compartment codes (m = mitochondria, c = cytosol) as well as metab__
prefix is added to HMDB IDs. More information regarding compartment
codes can be found <a href="http://bigg.ucsd.edu/compartments">here</a>.</p><pre class="language-r" tabindex="0"><code class="language-r">metab_to_HMDB <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>
  read_csv<span class="token punctuation">(</span><span class="token string">"data/metabolomic/MetaboliteToHMDB.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
metab_to_HMDB <span class="token operator">&lt;-</span> metab_to_HMDB<span class="token punctuation">[</span>metab_to_HMDB<span class="token operator">$</span>common <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
metab_inputs <span class="token operator">&lt;-</span> metab_inputs<span class="token punctuation">[</span>metab_to_HMDB<span class="token operator">$</span>common<span class="token punctuation">]</span>
names<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> metab_to_HMDB<span class="token operator">$</span>HMDB

metab_inputs <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span>prepare_metab_inputs<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">,</span> compartment_codes <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"m"</span><span class="token punctuation">,</span><span class="token string">"c"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>

metab_to_exclude <span class="token operator">&lt;-</span> metab_inputs<span class="token punctuation">[</span>abs<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;</span> <span class="token number">0.2</span><span class="token punctuation">]</span>
metab_inputs <span class="token operator">&lt;-</span> metab_inputs<span class="token punctuation">[</span>abs<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span> <span class="token operator">&gt;</span> <span class="token number">0.2</span><span class="token punctuation">]</span>
</code></pre><p>Next, the prior knowledge network (PKN) is loaded. To see how the full
meta PKN was assembled, see
<a href="https://github.com/saezlab/meta_PKN_BIGG.git">PKN</a>. Since we are not
interested in including transcription factors and metabolites that were
previously filtered out, we also remove the respective nodes from the
PKN.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">## Load and filter meta network</span>
<span class="token comment"># data("meta_network")</span>
<span class="token comment"># save(meta_network, file = "support/meta_network.RData")</span>
load<span class="token punctuation">(</span><span class="token string">"support/meta_network.RData"</span><span class="token punctuation">)</span>

meta_network <span class="token operator">&lt;-</span> meta_network_cleanup<span class="token punctuation">(</span>meta_network<span class="token punctuation">)</span>

meta_network <span class="token operator">&lt;-</span> meta_network<span class="token punctuation">[</span><span class="token operator">!</span><span class="token punctuation">(</span>meta_network<span class="token operator">$</span>source <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>TF_to_remove<span class="token punctuation">)</span><span class="token punctuation">)</span> 
                             <span class="token operator">&amp;</span> <span class="token operator">!</span><span class="token punctuation">(</span>meta_network<span class="token operator">$</span>target <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>TF_to_remove<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
meta_network <span class="token operator">&lt;-</span> meta_network<span class="token punctuation">[</span><span class="token operator">!</span><span class="token punctuation">(</span>meta_network<span class="token operator">$</span>source <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>metab_to_exclude<span class="token punctuation">)</span><span class="token punctuation">)</span> 
                             <span class="token operator">&amp;</span> <span class="token operator">!</span><span class="token punctuation">(</span>meta_network<span class="token operator">$</span>target <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>metab_to_exclude<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
</code></pre><p>Then the datasets are merged, entries that are not included in the PKN
are removed and the filtering steps are completed. The merging in this
case is based on the idea of deriving the forward network via the
receptor to transcription factor/metabolite direction. We will perform
first the first cosmos run between recptors and downstream TFs and
metabolites. The second run will focus on connecting TFs awith
downstream ligands.</p><p>Thus, we first define the upstream input as receptors, and the
downstream input as TFs and metabolites. Since the scoring of the
network is sensitive to the scale of the downstream measurments, we are
multiplying the metabolite weights by 10 so that they are on a similar
scale as the TF scores. Other scaling methods can be performed at the
user’S discretion, this one is just a rough scaling, but good enough in
this case.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># upstream_inputs &lt;- c(rec_inputs)</span>
<span class="token comment"># downstream_inputs &lt;- c(metab_inputs*10, signaling_input)</span>
<span class="token comment"># </span>
<span class="token comment"># upstream_inputs_filtered &lt;- cosmosR:::filter_input_nodes_not_in_pkn(upstream_inputs, meta_network)</span>
<span class="token comment"># downstream_inputs_filtered &lt;- cosmosR:::filter_input_nodes_not_in_pkn(downstream_inputs, meta_network)</span>
<span class="token comment"># </span>
<span class="token comment"># rec_to_TF_cosmos_input &lt;- list(upstream_inputs_filtered, downstream_inputs_filtered)</span>
<span class="token comment"># save(rec_to_TF_cosmos_input, file = "data/cosmos/rec_to_TF_cosmos_input.Rdata")</span>
</code></pre><h3 id="moon-meta-footprint-analysis">MOON (meta footprint analysis)</h3><p>In this part, we will run the network scoring to generate mechanistic
hypotheses connecting ligands, receptors, TFs and metabolites, while
pruning the network with information from both RNA and protein abundance
to constrain the flux of signal consistent with the molecular data
available.</p><p>In this approach, the network is compressed to avoid having redundant
paths in the network leading to inflating the importance of some input
measurements.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">##filter expressed genes from PKN</span>
<span class="token comment"># data("meta_network")</span>
<span class="token comment"># save(meta_network, file = "support/meta_network.RData")</span>
<span class="token comment"># load("support/meta_network.RData")</span>

<span class="token comment"># meta_network &lt;- meta_network_cleanup(meta_network)</span>

expressed_genes <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>read_csv<span class="token punctuation">(</span><span class="token string">"data/RNA/RNA_log2_FPKM_clean.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## Rows: 11265 Columns: 61</span>
<span class="token comment">## ── Column specification ────────────────────────────────────────────────────────</span>
<span class="token comment">## Delimiter: ","</span>
<span class="token comment">## chr  (1): Genes</span>
<span class="token comment">## dbl (60): 786-0, A498, A549/ATCC, ACHN, BT-549, CAKI-1, CCRF-CEM, COLO 205, ...</span>
<span class="token comment">## </span>
<span class="token comment">## ℹ Use `spec()` to retrieve the full column specification for this data.</span>
<span class="token comment">## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">expressed_genes <span class="token operator">&lt;-</span> setNames<span class="token punctuation">(</span>rep<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span>length<span class="token punctuation">(</span>expressed_genes<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">,</span> expressed_genes<span class="token operator">$</span>Genes<span class="token punctuation">)</span>
meta_network_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_pkn_expressed_genes<span class="token punctuation">(</span>names<span class="token punctuation">(</span>expressed_genes<span class="token punctuation">)</span><span class="token punctuation">,</span> meta_pkn <span class="token operator">=</span> meta_network<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing unexpressed nodes from PKN..."</span>
<span class="token comment">## [1] "COSMOS: 14430 interactions removed"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">##format metab inputs</span>
metab_inputs <span class="token operator">&lt;-</span> as.numeric<span class="token punctuation">(</span>scale<span class="token punctuation">(</span>weights<span class="token operator">$</span>metab<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">4</span><span class="token punctuation">]</span><span class="token punctuation">,</span> center <span class="token operator">=</span> F<span class="token punctuation">)</span><span class="token punctuation">)</span>
names<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> row.names<span class="token punctuation">(</span>weights<span class="token operator">$</span>metab<span class="token punctuation">)</span>

metab_to_HMDB <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>
  read_csv<span class="token punctuation">(</span><span class="token string">"data/metabolomic/MetaboliteToHMDB.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## Rows: 139 Columns: 2</span>
<span class="token comment">## ── Column specification ────────────────────────────────────────────────────────</span>
<span class="token comment">## Delimiter: ","</span>
<span class="token comment">## chr (2): common, HMDB</span>
<span class="token comment">## </span>
<span class="token comment">## ℹ Use `spec()` to retrieve the full column specification for this data.</span>
<span class="token comment">## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">metab_to_HMDB <span class="token operator">&lt;-</span> metab_to_HMDB<span class="token punctuation">[</span>metab_to_HMDB<span class="token operator">$</span>common <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token punctuation">]</span>
metab_inputs <span class="token operator">&lt;-</span> metab_inputs<span class="token punctuation">[</span>metab_to_HMDB<span class="token operator">$</span>common<span class="token punctuation">]</span>
names<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> metab_to_HMDB<span class="token operator">$</span>HMDB

metab_inputs <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span>prepare_metab_inputs<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">,</span> compartment_codes <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"m"</span><span class="token punctuation">,</span><span class="token string">"c"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "Adding compartment codes."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#prepare upstream inputs</span>
upstream_inputs <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span>rec_inputs<span class="token punctuation">)</span> <span class="token comment">#the upstream input should be filtered for most significant</span>

TF_inputs <span class="token operator">&lt;-</span> scale<span class="token punctuation">(</span>ligrec_TF_moon_inputs<span class="token operator">$</span>TF<span class="token punctuation">,</span> center <span class="token operator">=</span> F<span class="token punctuation">)</span>
TF_inputs <span class="token operator">&lt;-</span> setNames<span class="token punctuation">(</span>TF_inputs<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span> row.names<span class="token punctuation">(</span>TF_inputs<span class="token punctuation">)</span><span class="token punctuation">)</span>

downstream_inputs <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">,</span> TF_inputs<span class="token punctuation">)</span> <span class="token comment">#the downstream input should be complete</span>

upstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>upstream_inputs<span class="token punctuation">,</span> meta_network_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 5 input/measured nodes are not in PKN any more: BCAM, CD63, LRP10, NOTCH1, RXRA and 0 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">downstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>downstream_inputs<span class="token punctuation">,</span> meta_network_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 429 input/measured nodes are not in PKN any more: Metab__HMDB0011747_m, Metab__HMDB0000755_m, Metab__HMDB0001294_m, Metab__HMDB0001508_m, Metab__HMDB0000012_m, Metab__HMDB0001191_m and 423 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#Filter inputs and prune the meta_network to only keep nodes that can be found downstream of the inputs</span>
<span class="token comment">#The number of step is quite flexible, 7 steps already covers most of the network</span>

n_steps <span class="token operator">&lt;-</span> <span class="token number">6</span>

<span class="token comment"># in this step we prune the network to keep only the relevant part between upstream and downstream nodes</span>
meta_network_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>keep_controllable_neighbours<span class="token punctuation">(</span>meta_network_filtered
                                                       <span class="token punctuation">,</span> n_steps<span class="token punctuation">,</span> 
                                                       names<span class="token punctuation">(</span>upstream_inputs_filtered<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing nodes that are not reachable from inputs within 6 steps"</span>
<span class="token comment">## [1] "COSMOS: 29459 from  42695 interactions are removed from the PKN"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">downstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>downstream_inputs_filtered<span class="token punctuation">,</span> meta_network_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 28 input/measured nodes are not in PKN any more: Metab__HMDB0000168_m, Metab__HMDB0000056_m, Metab__HMDB0001051_m, Metab__HMDB0000139_m, Metab__HMDB0000687_m, Metab__HMDB0000696_m and 22 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">meta_network_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>keep_observable_neighbours<span class="token punctuation">(</span>meta_network_filtered<span class="token punctuation">,</span> n_steps<span class="token punctuation">,</span> names<span class="token punctuation">(</span>downstream_inputs_filtered<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing nodes that are not observable by measurements within 6 steps"</span>
<span class="token comment">## [1] "COSMOS: 6200 from  13236 interactions are removed from the PKN"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">upstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>upstream_inputs_filtered<span class="token punctuation">,</span> meta_network_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 4 input/measured nodes are not in PKN any more: EPHA6, GPC1, SDC1, SIRPA and 0 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">write_csv<span class="token punctuation">(</span>meta_network_filtered<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/meta_network_filtered.csv"</span><span class="token punctuation">)</span>

<span class="token comment">#compress the network to avoid redundant master controllers</span>
meta_network_compressed_list <span class="token operator">&lt;-</span> compress_same_children<span class="token punctuation">(</span>meta_network_filtered<span class="token punctuation">,</span> sig_input <span class="token operator">=</span> upstream_inputs_filtered<span class="token punctuation">,</span> metab_input <span class="token operator">=</span> downstream_inputs_filtered<span class="token punctuation">)</span>

meta_network_compressed <span class="token operator">&lt;-</span> meta_network_compressed_list<span class="token operator">$</span>compressed_network

meta_network_compressed <span class="token operator">&lt;-</span> meta_network_cleanup<span class="token punctuation">(</span>meta_network_compressed<span class="token punctuation">)</span>

load<span class="token punctuation">(</span>file <span class="token operator">=</span> <span class="token string">"support/dorothea_df.RData"</span><span class="token punctuation">)</span>

<span class="token comment"># RNA_input &lt;- as.numeric(weights$RNA[,4])</span>
<span class="token comment"># names(RNA_input) &lt;- gsub("_RNA$","",row.names(weights$RNA))</span>
</code></pre><p>In this step, we format the MOFA weight and activity score information
so that it can be appended to the moon network result afterward.</p><pre class="language-r" tabindex="0"><code class="language-r">data<span class="token punctuation">(</span><span class="token string">"HMDB_mapper_vec"</span><span class="token punctuation">)</span>

<span class="token comment">## MOFA weights</span>
MOFA_weights <span class="token operator">&lt;-</span> get_weights<span class="token punctuation">(</span>model<span class="token punctuation">,</span> factors <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">,</span> as.data.frame <span class="token operator">=</span> T<span class="token punctuation">)</span>
MOFA_weights <span class="token operator">&lt;-</span> MOFA_weights <span class="token percent-operator operator">%&gt;%</span>
  arrange<span class="token punctuation">(</span>feature<span class="token punctuation">,</span> view<span class="token punctuation">)</span> <span class="token percent-operator operator">%&gt;%</span>
  filter<span class="token punctuation">(</span><span class="token operator">!</span>duplicated<span class="token punctuation">(</span>feature<span class="token punctuation">)</span><span class="token punctuation">)</span>
MOFA_weights <span class="token operator">&lt;-</span> MOFA_weights<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
colnames<span class="token punctuation">(</span>MOFA_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"Nodes"</span><span class="token punctuation">,</span> <span class="token string">"mofa_weights"</span><span class="token punctuation">)</span>

<span class="token comment">#remove RNA and proteo tags and integrate values</span>
MOFA_weights<span class="token operator">$</span>Nodes <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_RNA$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>MOFA_weights<span class="token operator">$</span>Nodes<span class="token punctuation">)</span>
MOFA_weights<span class="token operator">$</span>Nodes <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_proteo$"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>MOFA_weights<span class="token operator">$</span>Nodes<span class="token punctuation">)</span>
MOFA_weights<span class="token operator">$</span>sign <span class="token operator">&lt;-</span> sign<span class="token punctuation">(</span>MOFA_weights<span class="token operator">$</span>mofa_weights<span class="token punctuation">)</span>
MOFA_weights<span class="token operator">$</span>mofa_weights <span class="token operator">&lt;-</span> abs<span class="token punctuation">(</span>MOFA_weights<span class="token operator">$</span>mofa_weights<span class="token punctuation">)</span>

MOFA_weights <span class="token operator">&lt;-</span> MOFA_weights <span class="token percent-operator operator">%&gt;%</span> group_by<span class="token punctuation">(</span>Nodes<span class="token punctuation">)</span> <span class="token percent-operator operator">%&gt;%</span> summarise_each<span class="token punctuation">(</span>funs<span class="token punctuation">(</span>max<span class="token punctuation">(</span>.<span class="token punctuation">,</span> na.rm <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## Warning: `summarise_each()` was deprecated in dplyr 0.7.0.</span>
<span class="token comment">## ℹ Please use `across()` instead.</span>
<span class="token comment">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span>
<span class="token comment">## generated.</span>

<span class="token comment">## Warning: `funs()` was deprecated in dplyr 0.8.0.</span>
<span class="token comment">## ℹ Please use a list of either functions or lambdas:</span>
<span class="token comment">## </span>
<span class="token comment">## # Simple named list: list(mean = mean, median = median)</span>
<span class="token comment">## </span>
<span class="token comment">## # Auto named with `tibble::lst()`: tibble::lst(mean, median)</span>
<span class="token comment">## </span>
<span class="token comment">## # Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))</span>
<span class="token comment">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span>
<span class="token comment">## generated.</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">MOFA_weights <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>MOFA_weights<span class="token punctuation">)</span>
MOFA_weights<span class="token operator">$</span>mofa_weights <span class="token operator">&lt;-</span> MOFA_weights<span class="token operator">$</span>mofa_weights <span class="token operator">*</span> MOFA_weights<span class="token operator">$</span>sign
MOFA_weights <span class="token operator">&lt;-</span> MOFA_weights<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token operator">-</span><span class="token number">3</span><span class="token punctuation">]</span>

<span class="token comment">#make metabolite names coherent between mofa and cosmos</span>
<span class="token comment">#for now i just add the metab input that is already formatted</span>
<span class="token comment">#I will miss some weights that were not input but will fix that at later time</span>
MOFA_weights_metabaddon <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>metab_inputs<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>MOFA_weights_metabaddon<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> <span class="token string">"mofa_weights"</span>
MOFA_weights_metabaddon<span class="token operator">$</span>Nodes <span class="token operator">&lt;-</span> row.names<span class="token punctuation">(</span>MOFA_weights_metabaddon<span class="token punctuation">)</span>

MOFA_weights_metabaddon<span class="token punctuation">[</span><span class="token punctuation">,</span> <span class="token number">2</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> sapply<span class="token punctuation">(</span>MOFA_weights_metabaddon<span class="token punctuation">[</span><span class="token punctuation">,</span> <span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>x<span class="token punctuation">,</span> HMDB_mapper_vec<span class="token punctuation">)</span> <span class="token punctuation">{</span>
        x <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"Metab__"</span><span class="token punctuation">,</span> <span class="token string">""</span><span class="token punctuation">,</span> x<span class="token punctuation">)</span>
        suffixe <span class="token operator">&lt;-</span> stringr<span class="token operator">::</span>str_extract<span class="token punctuation">(</span>x<span class="token punctuation">,</span> <span class="token string">"_[a-z]$"</span><span class="token punctuation">)</span>
        x <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_[a-z]$"</span><span class="token punctuation">,</span> <span class="token string">""</span><span class="token punctuation">,</span> x<span class="token punctuation">)</span>
        <span class="token keyword">if</span> <span class="token punctuation">(</span>x <span class="token percent-operator operator">%in%</span> names<span class="token punctuation">(</span>HMDB_mapper_vec<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
            x <span class="token operator">&lt;-</span> HMDB_mapper_vec<span class="token punctuation">[</span>x<span class="token punctuation">]</span>
            x <span class="token operator">&lt;-</span> paste<span class="token punctuation">(</span><span class="token string">"Metab__"</span><span class="token punctuation">,</span> x<span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span>
        <span class="token punctuation">}</span>
        <span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>is.na<span class="token punctuation">(</span>suffixe<span class="token punctuation">)</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
            x <span class="token operator">&lt;-</span> paste<span class="token punctuation">(</span>x<span class="token punctuation">,</span> suffixe<span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span>
        <span class="token punctuation">}</span>
        return<span class="token punctuation">(</span>x<span class="token punctuation">)</span>
    <span class="token punctuation">}</span><span class="token punctuation">,</span> HMDB_mapper_vec <span class="token operator">=</span> HMDB_mapper_vec<span class="token punctuation">)</span>

MOFA_weights <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>rbind<span class="token punctuation">(</span>MOFA_weights<span class="token punctuation">,</span> MOFA_weights_metabaddon<span class="token punctuation">)</span><span class="token punctuation">)</span>


<span class="token comment">## Ligands</span>
lig_weights <span class="token operator">&lt;-</span> ligrec_TF_moon_inputs<span class="token operator">$</span>ligrec
names<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"___.+"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span><span class="token punctuation">)</span>
lig_weights <span class="token operator">&lt;-</span> tapply<span class="token punctuation">(</span>lig_weights<span class="token punctuation">,</span> names<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span><span class="token punctuation">,</span> mean<span class="token punctuation">)</span>
ligands <span class="token operator">&lt;-</span> names<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span>
lig_weights <span class="token operator">&lt;-</span> as.numeric<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> ligands
lig_weights <span class="token operator">&lt;-</span> data.frame<span class="token punctuation">(</span>Nodes <span class="token operator">=</span> names<span class="token punctuation">(</span>lig_weights<span class="token punctuation">)</span><span class="token punctuation">,</span> feature_weights <span class="token operator">=</span> lig_weights<span class="token punctuation">)</span>

<span class="token comment">## Receptors</span>
rec_weights <span class="token operator">&lt;-</span> ligrec_TF_moon_inputs<span class="token operator">$</span>ligrec
names<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">".+___"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span><span class="token punctuation">)</span>
rec_weights <span class="token operator">&lt;-</span> tapply<span class="token punctuation">(</span>rec_weights<span class="token punctuation">,</span> names<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span><span class="token punctuation">,</span> mean<span class="token punctuation">)</span>
receptors <span class="token operator">&lt;-</span> names<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span>
rec_weights <span class="token operator">&lt;-</span> as.numeric<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> receptors
rec_weights <span class="token operator">&lt;-</span> data.frame<span class="token punctuation">(</span>Nodes <span class="token operator">=</span> names<span class="token punctuation">(</span>rec_weights<span class="token punctuation">)</span><span class="token punctuation">,</span> feature_weights <span class="token operator">=</span> rec_weights<span class="token punctuation">)</span>

<span class="token comment">## TF</span>
TF_weights <span class="token operator">&lt;-</span> ligrec_TF_moon_inputs<span class="token operator">$</span>TF
names<span class="token punctuation">(</span>TF_weights<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"_TF"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>TF_weights<span class="token punctuation">)</span><span class="token punctuation">)</span>
TF_weights <span class="token operator">&lt;-</span> TF_weights<span class="token punctuation">[</span><span class="token operator">!</span><span class="token punctuation">(</span>names<span class="token punctuation">(</span>TF_weights<span class="token punctuation">)</span> <span class="token percent-operator operator">%in%</span> c<span class="token punctuation">(</span>rec_weights<span class="token operator">$</span>Nodes<span class="token punctuation">,</span>lig_weights<span class="token operator">$</span>Nodes<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
TF_weights <span class="token operator">&lt;-</span> data.frame<span class="token punctuation">(</span>Nodes <span class="token operator">=</span> names<span class="token punctuation">(</span>TF_weights<span class="token punctuation">)</span><span class="token punctuation">,</span> feature_weights <span class="token operator">=</span> TF_weights<span class="token punctuation">)</span>



<span class="token comment">## Combine</span>
feature_weights <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>rbind<span class="token punctuation">(</span>TF_weights<span class="token punctuation">,</span> rec_weights<span class="token punctuation">,</span> lig_weights<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><p>To also identify whether a node is a ligand or receptor we also load the
liana consensus LR ressource.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># LIANA</span>
load<span class="token punctuation">(</span><span class="token string">"support/ligrec_ressource.RData"</span><span class="token punctuation">)</span>
ligrec_df <span class="token operator">&lt;-</span> ligrec_ressource<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token string">"source_genesymbol"</span><span class="token punctuation">,</span><span class="token string">"target_genesymbol"</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
ligrec_df <span class="token operator">&lt;-</span> distinct<span class="token punctuation">(</span>ligrec_df<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>ligrec_df<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"Node1"</span><span class="token punctuation">,</span><span class="token string">"Node2"</span><span class="token punctuation">)</span>

ligrec_df<span class="token operator">$</span>Node1 <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"-"</span><span class="token punctuation">,</span><span class="token string">"_"</span><span class="token punctuation">,</span>ligrec_df<span class="token operator">$</span>Node1<span class="token punctuation">)</span>
ligrec_df<span class="token operator">$</span>Node2 <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"-"</span><span class="token punctuation">,</span><span class="token string">"_"</span><span class="token punctuation">,</span>ligrec_df<span class="token operator">$</span>Node2<span class="token punctuation">)</span>
ligrec_df<span class="token operator">$</span>Sign <span class="token operator">&lt;-</span> <span class="token number">1</span>
ligrec_df<span class="token operator">$</span>Weight <span class="token operator">&lt;-</span> <span class="token number">1</span>
</code></pre><p>In this step. we run the first moon analysis, to connect receptors with
downstream TFs and ligands. MOON function is run iterativelly in a loop
until the resulting scored network does not include edges incoherent
with the RNA and proteomic data measurents. then, the network is
decompressed, IDs are translated in a human-friendly format and a
subnetowrk comparable to the classic cosmos solution network is
generated with the reduce_solution_network function.</p><pre class="language-r" tabindex="0"><code class="language-r">meta_network_rec_to_TFmetab <span class="token operator">&lt;-</span> meta_network_compressed

<span class="token comment">#We run moon in a loop until TF-target coherence convergences</span>
before <span class="token operator">&lt;-</span> <span class="token number">1</span>
after <span class="token operator">&lt;-</span> <span class="token number">0</span>
i <span class="token operator">&lt;-</span> <span class="token number">1</span>
<span class="token keyword">while</span> <span class="token punctuation">(</span>before <span class="token operator">!=</span> after <span class="token operator">&amp;</span> i <span class="token operator">&lt;</span> <span class="token number">10</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
  before <span class="token operator">&lt;-</span> length<span class="token punctuation">(</span>meta_network_rec_to_TFmetab<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
  start_time <span class="token operator">&lt;-</span> Sys.time<span class="token punctuation">(</span><span class="token punctuation">)</span>
  moon_res <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span>moon<span class="token punctuation">(</span>upstream_input <span class="token operator">=</span> upstream_inputs_filtered<span class="token punctuation">,</span> 
                                                 downstream_input <span class="token operator">=</span> downstream_inputs_filtered<span class="token punctuation">,</span> 
                                                 meta_network <span class="token operator">=</span> meta_network_rec_to_TFmetab<span class="token punctuation">,</span> 
                                                 n_layers <span class="token operator">=</span> n_steps<span class="token punctuation">,</span> 
                                                 statistic <span class="token operator">=</span> <span class="token string">"ulm"</span><span class="token punctuation">)</span> 
  
  meta_network_rec_to_TFmetab <span class="token operator">&lt;-</span> filter_incohrent_TF_target<span class="token punctuation">(</span>moon_res<span class="token punctuation">,</span> dorothea_df<span class="token punctuation">,</span> meta_network_rec_to_TFmetab<span class="token punctuation">,</span> RNA_input<span class="token punctuation">)</span>
  end_time <span class="token operator">&lt;-</span> Sys.time<span class="token punctuation">(</span><span class="token punctuation">)</span>
  
  print<span class="token punctuation">(</span>end_time <span class="token operator">-</span> start_time<span class="token punctuation">)</span>
  
  after <span class="token operator">&lt;-</span> length<span class="token punctuation">(</span>meta_network_rec_to_TFmetab<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
  i <span class="token operator">&lt;-</span> i <span class="token operator">+</span> <span class="token number">1</span>
<span class="token punctuation">}</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] 2</span>
<span class="token comment">## [1] 3</span>
<span class="token comment">## [1] 4</span>
<span class="token comment">## [1] 5</span>
<span class="token comment">## [1] 6</span>
<span class="token comment">## Time difference of 0.1666861 secs</span>
<span class="token comment">## [1] 2</span>
<span class="token comment">## [1] 3</span>
<span class="token comment">## [1] 4</span>
<span class="token comment">## [1] 5</span>
<span class="token comment">## [1] 6</span>
<span class="token comment">## Time difference of 0.2143788 secs</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token keyword">if</span><span class="token punctuation">(</span>i <span class="token operator">&lt;</span> <span class="token number">10</span><span class="token punctuation">)</span>
<span class="token punctuation">{</span>
  print<span class="token punctuation">(</span>paste<span class="token punctuation">(</span><span class="token string">"Converged after "</span><span class="token punctuation">,</span>paste<span class="token punctuation">(</span>i<span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token string">" iterations"</span><span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">,</span>sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span> <span class="token keyword">else</span>
<span class="token punctuation">{</span>
  print<span class="token punctuation">(</span>paste<span class="token punctuation">(</span><span class="token string">"Interupted after "</span><span class="token punctuation">,</span>paste<span class="token punctuation">(</span>i<span class="token punctuation">,</span><span class="token string">" iterations. Convergence uncertain."</span><span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">,</span>sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "Converged after 2 iterations"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">moon_res <span class="token operator">&lt;-</span> decompress_moon_result<span class="token punctuation">(</span>moon_res<span class="token punctuation">,</span> meta_network_compressed_list<span class="token punctuation">,</span> meta_network_rec_to_TFmetab<span class="token punctuation">)</span>

<span class="token comment">#save the whole res for later</span>
moon_res_rec_to_TFmet <span class="token operator">&lt;-</span> moon_res<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
moon_res_rec_to_TFmet<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> translate_column_HMDB<span class="token punctuation">(</span>moon_res_rec_to_TFmet<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span> HMDB_mapper_vec<span class="token punctuation">)</span>

levels <span class="token operator">&lt;-</span> moon_res<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>

moon_res <span class="token operator">&lt;-</span> moon_res<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">4</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
names<span class="token punctuation">(</span>moon_res<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"source"</span>

plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>moon_res<span class="token operator">$</span>score<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/Run%20moon%20rec_to_TFmetab-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">solution_network <span class="token operator">&lt;-</span> reduce_solution_network<span class="token punctuation">(</span>decoupleRnival_res <span class="token operator">=</span> moon_res<span class="token punctuation">,</span> 
                                            meta_network <span class="token operator">=</span> meta_network_filtered<span class="token punctuation">,</span>
                                            cutoff <span class="token operator">=</span> <span class="token number">1.5</span><span class="token punctuation">,</span> 
                                            upstream_input <span class="token operator">=</span> upstream_inputs_filtered<span class="token punctuation">,</span> 
                                            RNA_input <span class="token operator">=</span> RNA_input<span class="token punctuation">,</span> 
                                            n_steps <span class="token operator">=</span> n_steps<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing nodes that are not reachable from inputs within 6 steps"</span>
<span class="token comment">## [1] "COSMOS: 385 from  901 interactions are removed from the PKN"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">SIF_rec_to_TFmetab <span class="token operator">&lt;-</span> solution_network<span class="token operator">$</span>SIF
names<span class="token punctuation">(</span>SIF_rec_to_TFmetab<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">3</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"sign"</span>
ATT_rec_to_TFmetab <span class="token operator">&lt;-</span> solution_network<span class="token operator">$</span>ATT

translated_res <span class="token operator">&lt;-</span> translate_res<span class="token punctuation">(</span>SIF_rec_to_TFmetab<span class="token punctuation">,</span>ATT_rec_to_TFmetab<span class="token punctuation">,</span>HMDB_mapper_vec<span class="token punctuation">)</span>

levels_translated <span class="token operator">&lt;-</span> translate_res<span class="token punctuation">(</span>SIF_rec_to_TFmetab<span class="token punctuation">,</span>levels<span class="token punctuation">,</span>HMDB_mapper_vec<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">]</span>

SIF_rec_to_TFmetab <span class="token operator">&lt;-</span> translated_res<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span>
ATT_rec_to_TFmetab <span class="token operator">&lt;-</span> translated_res<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">]</span>

<span class="token comment">## Add weights to data</span>
ATT_rec_to_TFmetab <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token punctuation">,</span> MOFA_weights<span class="token punctuation">,</span> all.x <span class="token operator">=</span> T<span class="token punctuation">)</span>
ATT_rec_to_TFmetab <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token punctuation">,</span> feature_weights<span class="token punctuation">,</span> all.x <span class="token operator">=</span> T<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"AvgAct"</span>

ATT_rec_to_TFmetab<span class="token operator">$</span>NodeType <span class="token operator">&lt;-</span> ifelse<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> levels_translated<span class="token punctuation">[</span>levels_translated<span class="token operator">$</span>level <span class="token operator">==</span> <span class="token number">0</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">0</span><span class="token punctuation">)</span>

ATT_rec_to_TFmetab<span class="token operator">$</span>NodeType <span class="token operator">&lt;-</span> ifelse<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> ligrec_df<span class="token operator">$</span>Node1<span class="token punctuation">,</span> <span class="token number">2</span><span class="token punctuation">,</span> ifelse<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> ligrec_df<span class="token operator">$</span>Node2<span class="token punctuation">,</span> <span class="token number">3</span><span class="token punctuation">,</span> ifelse<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> dorothea_df<span class="token operator">$</span>source<span class="token punctuation">,</span> <span class="token number">4</span><span class="token punctuation">,</span> ATT_rec_to_TFmetab<span class="token operator">$</span>NodeType<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span> 

names<span class="token punctuation">(</span>SIF_rec_to_TFmetab<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">4</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"Weight"</span>

write_csv<span class="token punctuation">(</span>SIF_rec_to_TFmetab<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/SIF_rec_TFmetab.csv"</span><span class="token punctuation">)</span>
write_csv<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/ATT_rec_TFmetab.csv"</span><span class="token punctuation">)</span>
</code></pre><p>Next, we also score a network connecting the TF with downstream ligands.
The procedure is the same as the previous section, only the upstream and
downstream input definition is changing.</p><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">##filter expressed genes from PKN</span>
dorothea_PKN <span class="token operator">&lt;-</span> dorothea_df<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
names<span class="token punctuation">(</span>dorothea_PKN<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"interaction"</span>

dorothea_PKN_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_pkn_expressed_genes<span class="token punctuation">(</span>names<span class="token punctuation">(</span>expressed_genes<span class="token punctuation">)</span><span class="token punctuation">,</span> meta_pkn <span class="token operator">=</span> dorothea_PKN<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing unexpressed nodes from PKN..."</span>
<span class="token comment">## [1] "COSMOS: 12053 interactions removed"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#prepare upstream inputs</span>

<span class="token comment">#the upstream input should be filtered for most significant</span>
upstream_inputs <span class="token operator">&lt;-</span> setNames<span class="token punctuation">(</span>TF_weights<span class="token operator">$</span>feature_weights<span class="token punctuation">,</span> TF_weights<span class="token operator">$</span>Nodes<span class="token punctuation">)</span>
upstream_inputs <span class="token operator">&lt;-</span> upstream_inputs<span class="token punctuation">[</span>abs<span class="token punctuation">(</span>upstream_inputs<span class="token punctuation">)</span> <span class="token operator">&gt;</span> <span class="token number">2</span><span class="token punctuation">]</span>

lig_inputs <span class="token operator">&lt;-</span> ligrec_TF_moon_inputs<span class="token operator">$</span>ligrec
names<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> gsub<span class="token punctuation">(</span><span class="token string">"___.+"</span><span class="token punctuation">,</span><span class="token string">""</span><span class="token punctuation">,</span>names<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span><span class="token punctuation">)</span>
lig_inputs <span class="token operator">&lt;-</span> tapply<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">,</span> names<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span><span class="token punctuation">,</span> mean<span class="token punctuation">)</span>
ligands <span class="token operator">&lt;-</span> names<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span>
lig_inputs <span class="token operator">&lt;-</span> as.numeric<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span> <span class="token operator">&lt;-</span> ligands

lig_inputs <span class="token operator">&lt;-</span> scale<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">,</span> center <span class="token operator">=</span> F<span class="token punctuation">)</span>
lig_inputs <span class="token operator">&lt;-</span> setNames<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span> row.names<span class="token punctuation">(</span>lig_inputs<span class="token punctuation">)</span><span class="token punctuation">)</span>

downstream_inputs <span class="token operator">&lt;-</span> lig_inputs <span class="token comment">#the downstream input should be complete</span>

upstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>upstream_inputs<span class="token punctuation">,</span> dorothea_PKN_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 22 input/measured nodes are not in PKN any more: SPI1, NR0B2, ESR1, AR, NKX2-5, RARB and 16 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">downstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>downstream_inputs<span class="token punctuation">,</span> dorothea_PKN_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 19 input/measured nodes are not in PKN any more: ACTR2, ADAM9, AGRN, ARPC5, EFNA3, EFNB1 and 13 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment">#Filter inputs and prune the meta_network to only keep nodes that can be found downstream of the inputs</span>
<span class="token comment">#The number of step is quite flexible, 7 steps already covers most of the network</span>

n_steps <span class="token operator">&lt;-</span> <span class="token number">1</span>

<span class="token comment"># in this step we prune the network to keep only the relevant part between upstream and downstream nodes</span>
dorothea_PKN_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>keep_controllable_neighbours<span class="token punctuation">(</span>dorothea_PKN_filtered
                                                       <span class="token punctuation">,</span> n_steps<span class="token punctuation">,</span> 
                                                       names<span class="token punctuation">(</span>upstream_inputs_filtered<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing nodes that are not reachable from inputs within 1 steps"</span>
<span class="token comment">## [1] "COSMOS: 1896 from  17542 interactions are removed from the PKN"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">downstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>downstream_inputs_filtered<span class="token punctuation">,</span> dorothea_PKN_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 9 input/measured nodes are not in PKN any more: BMP1, ETV5, GAS6, GDF11, ITGB3BP, MAML2 and 3 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">dorothea_PKN_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>keep_observable_neighbours<span class="token punctuation">(</span>dorothea_PKN_filtered<span class="token punctuation">,</span> n_steps<span class="token punctuation">,</span> names<span class="token punctuation">(</span>downstream_inputs_filtered<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing nodes that are not observable by measurements within 1 steps"</span>
<span class="token comment">## [1] "COSMOS: 12699 from  15646 interactions are removed from the PKN"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">upstream_inputs_filtered <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span><span class="token operator">:</span>filter_input_nodes_not_in_pkn<span class="token punctuation">(</span>upstream_inputs_filtered<span class="token punctuation">,</span> dorothea_PKN_filtered<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: 7 input/measured nodes are not in PKN any more: MTF1, NR2C2, RBPJ, SRF, NCOA2, E2F2 and 1 more."</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r"><span class="token comment"># load(file = "support/dorothea_df.RData")</span>

<span class="token comment"># RNA_input &lt;- as.numeric(weights$RNA[,4])</span>
<span class="token comment"># names(RNA_input) &lt;- gsub("_RNA$","",row.names(weights$RNA))</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">meta_network_TF_lig <span class="token operator">&lt;-</span> dorothea_PKN_filtered

write_csv<span class="token punctuation">(</span>meta_network_TF_lig<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/meta_network_TF_lig.csv"</span><span class="token punctuation">)</span>

before <span class="token operator">&lt;-</span> <span class="token number">1</span>
after <span class="token operator">&lt;-</span> <span class="token number">0</span>
i <span class="token operator">&lt;-</span> <span class="token number">1</span>
<span class="token keyword">while</span> <span class="token punctuation">(</span>before <span class="token operator">!=</span> after <span class="token operator">&amp;</span> i <span class="token operator">&lt;</span> <span class="token number">10</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
  before <span class="token operator">&lt;-</span> length<span class="token punctuation">(</span>meta_network_TF_lig<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
  moon_res <span class="token operator">&lt;-</span> cosmosR<span class="token operator">::</span>moon<span class="token punctuation">(</span>upstream_input <span class="token operator">=</span> upstream_inputs_filtered<span class="token punctuation">,</span> 
                                                 downstream_input <span class="token operator">=</span> downstream_inputs_filtered<span class="token punctuation">,</span> 
                                                 meta_network <span class="token operator">=</span> meta_network_TF_lig<span class="token punctuation">,</span> 
                                                 n_layers <span class="token operator">=</span> n_steps<span class="token punctuation">,</span> 
                                                 statistic <span class="token operator">=</span> <span class="token string">"ulm"</span><span class="token punctuation">)</span> 
  
  meta_network_TF_lig <span class="token operator">&lt;-</span> filter_incohrent_TF_target<span class="token punctuation">(</span>moon_res<span class="token punctuation">,</span> dorothea_df<span class="token punctuation">,</span> meta_network_TF_lig<span class="token punctuation">,</span> RNA_input<span class="token punctuation">)</span>
  after <span class="token operator">&lt;-</span> length<span class="token punctuation">(</span>meta_network_TF_lig<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">)</span>
  i <span class="token operator">&lt;-</span> i <span class="token operator">+</span> <span class="token number">1</span>
<span class="token punctuation">}</span>

<span class="token keyword">if</span><span class="token punctuation">(</span>i <span class="token operator">&lt;</span> <span class="token number">10</span><span class="token punctuation">)</span>
<span class="token punctuation">{</span>
  print<span class="token punctuation">(</span>paste<span class="token punctuation">(</span><span class="token string">"Converged after "</span><span class="token punctuation">,</span>paste<span class="token punctuation">(</span>i<span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token string">" iterations"</span><span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">,</span>sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span> <span class="token keyword">else</span>
<span class="token punctuation">{</span>
  print<span class="token punctuation">(</span>paste<span class="token punctuation">(</span><span class="token string">"Interupted after "</span><span class="token punctuation">,</span>paste<span class="token punctuation">(</span>i<span class="token punctuation">,</span><span class="token string">" iterations. Convergence uncertain."</span><span class="token punctuation">,</span> sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">,</span>sep <span class="token operator">=</span> <span class="token string">""</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token punctuation">}</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "Converged after 1 iterations"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">moon_res_TF_lig <span class="token operator">&lt;-</span> moon_res
moon_res_TF_lig<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> translate_column_HMDB<span class="token punctuation">(</span>moon_res_TF_lig<span class="token punctuation">[</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span> HMDB_mapper_vec<span class="token punctuation">)</span>

levels <span class="token operator">&lt;-</span> moon_res<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">]</span>

moon_res <span class="token operator">&lt;-</span> moon_res<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
names<span class="token punctuation">(</span>moon_res<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"source"</span>

plot<span class="token punctuation">(</span>density<span class="token punctuation">(</span>moon_res<span class="token operator">$</span>score<span class="token punctuation">)</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">)</span>
abline<span class="token punctuation">(</span>v <span class="token operator">=</span> <span class="token operator">-</span><span class="token number">1</span><span class="token punctuation">)</span>
</code></pre><p><img src="https://raw.githubusercontent.com/saezlab/Factor_COSMOS/main/MOFA_to_COSMOS_files/figure-gfm/run%20moon%20TF%20to%20lig-1.png" alt=""><!-- --></p><pre class="language-r" tabindex="0"><code class="language-r">solution_network <span class="token operator">&lt;-</span> reduce_solution_network<span class="token punctuation">(</span>decoupleRnival_res <span class="token operator">=</span> moon_res<span class="token punctuation">,</span> 
                                            meta_network <span class="token operator">=</span> as.data.frame<span class="token punctuation">(</span>dorothea_PKN_filtered<span class="token punctuation">[</span><span class="token punctuation">,</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">]</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
                                            cutoff <span class="token operator">=</span> <span class="token number">1.5</span><span class="token punctuation">,</span> 
                                            upstream_input <span class="token operator">=</span> upstream_inputs_filtered<span class="token punctuation">,</span> 
                                            RNA_input <span class="token operator">=</span> RNA_input<span class="token punctuation">,</span> 
                                            n_steps <span class="token operator">=</span> n_steps<span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## [1] "COSMOS: removing nodes that are not reachable from inputs within 1 steps"</span>
<span class="token comment">## [1] "COSMOS: 29 from  143 interactions are removed from the PKN"</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">SIF_TF_lig <span class="token operator">&lt;-</span> solution_network<span class="token operator">$</span>SIF
names<span class="token punctuation">(</span>SIF_TF_lig<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">3</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"sign"</span>
ATT_TF_lig <span class="token operator">&lt;-</span> solution_network<span class="token operator">$</span>ATT


translated_res <span class="token operator">&lt;-</span> translate_res<span class="token punctuation">(</span>SIF_TF_lig<span class="token punctuation">,</span>ATT_TF_lig<span class="token punctuation">,</span>HMDB_mapper_vec<span class="token punctuation">)</span>

levels_translated <span class="token operator">&lt;-</span> translate_res<span class="token punctuation">(</span>SIF_TF_lig<span class="token punctuation">,</span>levels<span class="token punctuation">,</span>HMDB_mapper_vec<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">]</span>

SIF_TF_lig <span class="token operator">&lt;-</span> translated_res<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">]</span>
ATT_TF_lig <span class="token operator">&lt;-</span> translated_res<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">]</span>

<span class="token comment">## Add weights to data</span>
ATT_TF_lig <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>ATT_TF_lig<span class="token punctuation">,</span> MOFA_weights<span class="token punctuation">,</span> all.x <span class="token operator">=</span> T<span class="token punctuation">)</span>
ATT_TF_lig <span class="token operator">&lt;-</span> merge<span class="token punctuation">(</span>ATT_TF_lig<span class="token punctuation">,</span> feature_weights<span class="token punctuation">,</span> all.x <span class="token operator">=</span> T<span class="token punctuation">)</span>
names<span class="token punctuation">(</span>ATT_TF_lig<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"AvgAct"</span>

ATT_TF_lig<span class="token operator">$</span>NodeType <span class="token operator">&lt;-</span> ifelse<span class="token punctuation">(</span>ATT_TF_lig<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> levels_translated<span class="token punctuation">[</span>levels_translated<span class="token operator">$</span>level <span class="token operator">==</span> <span class="token number">0</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">0</span><span class="token punctuation">)</span>

ATT_TF_lig<span class="token operator">$</span>NodeType <span class="token operator">&lt;-</span> ifelse<span class="token punctuation">(</span>ATT_TF_lig<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> ligrec_df<span class="token operator">$</span>Node1<span class="token punctuation">,</span> <span class="token number">2</span><span class="token punctuation">,</span> ifelse<span class="token punctuation">(</span>ATT_TF_lig<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> ligrec_df<span class="token operator">$</span>Node2<span class="token punctuation">,</span> <span class="token number">3</span><span class="token punctuation">,</span> ifelse<span class="token punctuation">(</span>ATT_TF_lig<span class="token operator">$</span>Nodes <span class="token percent-operator operator">%in%</span> dorothea_df<span class="token operator">$</span>source<span class="token punctuation">,</span> <span class="token number">4</span><span class="token punctuation">,</span> ATT_TF_lig<span class="token operator">$</span>NodeType<span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span> 

names<span class="token punctuation">(</span>SIF_TF_lig<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">4</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> <span class="token string">"Weight"</span>

write_csv<span class="token punctuation">(</span>SIF_TF_lig<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/SIF_TF_lig.csv"</span><span class="token punctuation">)</span>
write_csv<span class="token punctuation">(</span>ATT_TF_lig<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/ATT_TF_lig.csv"</span><span class="token punctuation">)</span>
</code></pre><p>Both networks (Receptos -&gt; TF and metabs; TF -&gt; Ligands) are merged
together into a single compbined scored network. We also add the
conections between ligands and their corresponding receptors.</p><pre class="language-r" tabindex="0"><code class="language-r">combined_SIF_moon <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>rbind<span class="token punctuation">(</span>SIF_rec_to_TFmetab<span class="token punctuation">,</span> SIF_TF_lig<span class="token punctuation">)</span><span class="token punctuation">)</span>
combined_SIF_moon <span class="token operator">&lt;-</span> unique<span class="token punctuation">(</span>combined_SIF_moon<span class="token punctuation">)</span>

combined_ATT_moon <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>rbind<span class="token punctuation">(</span>ATT_rec_to_TFmetab<span class="token punctuation">,</span> ATT_TF_lig<span class="token punctuation">)</span><span class="token punctuation">)</span>
combined_ATT_moon <span class="token operator">&lt;-</span> combined_ATT_moon <span class="token percent-operator operator">%&gt;%</span> group_by<span class="token punctuation">(</span>Nodes<span class="token punctuation">)</span> <span class="token percent-operator operator">%&gt;%</span> summarise_each<span class="token punctuation">(</span>funs<span class="token punctuation">(</span>mean<span class="token punctuation">(</span>.<span class="token punctuation">,</span> na.rm <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## Warning: `summarise_each()` was deprecated in dplyr 0.7.0.</span>
<span class="token comment">## ℹ Please use `across()` instead.</span>
<span class="token comment">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span>
<span class="token comment">## generated.</span>

<span class="token comment">## Warning: `funs()` was deprecated in dplyr 0.8.0.</span>
<span class="token comment">## ℹ Please use a list of either functions or lambdas:</span>
<span class="token comment">## </span>
<span class="token comment">## # Simple named list: list(mean = mean, median = median)</span>
<span class="token comment">## </span>
<span class="token comment">## # Auto named with `tibble::lst()`: tibble::lst(mean, median)</span>
<span class="token comment">## </span>
<span class="token comment">## # Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))</span>
<span class="token comment">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span>
<span class="token comment">## generated.</span>
</code></pre><pre class="language-r" tabindex="0"><code class="language-r">combined_ATT_moon <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>combined_ATT_moon<span class="token punctuation">)</span>

ligrec_ressource_addon <span class="token operator">&lt;-</span> ligrec_ressource<span class="token punctuation">[</span>
  ligrec_ressource<span class="token operator">$</span>source_genesymbol <span class="token percent-operator operator">%in%</span> combined_SIF_moon<span class="token operator">$</span>target <span class="token operator">&amp;</span>
    ligrec_ressource<span class="token operator">$</span>target_genesymbol <span class="token percent-operator operator">%in%</span> combined_SIF_moon<span class="token operator">$</span>source
<span class="token punctuation">,</span> c<span class="token punctuation">(</span><span class="token number">10</span><span class="token punctuation">,</span><span class="token number">12</span><span class="token punctuation">)</span><span class="token punctuation">]</span>
ligrec_ressource_addon<span class="token operator">$</span>sign <span class="token operator">&lt;-</span> <span class="token number">1</span>
ligrec_ressource_addon<span class="token operator">$</span>Weight <span class="token operator">&lt;-</span> <span class="token number">1</span>
names<span class="token punctuation">(</span>ligrec_ressource_addon<span class="token punctuation">)</span><span class="token punctuation">[</span>c<span class="token punctuation">(</span><span class="token number">1</span><span class="token punctuation">,</span><span class="token number">2</span><span class="token punctuation">)</span><span class="token punctuation">]</span> <span class="token operator">&lt;-</span> c<span class="token punctuation">(</span><span class="token string">"source"</span><span class="token punctuation">,</span><span class="token string">"target"</span><span class="token punctuation">)</span>
ligrec_ressource_addon <span class="token operator">&lt;-</span> unique<span class="token punctuation">(</span>ligrec_ressource_addon<span class="token punctuation">)</span>

combined_SIF_moon <span class="token operator">&lt;-</span> as.data.frame<span class="token punctuation">(</span>rbind<span class="token punctuation">(</span>combined_SIF_moon<span class="token punctuation">,</span> ligrec_ressource_addon<span class="token punctuation">)</span><span class="token punctuation">)</span>

<span class="token comment">#It appears I may need to only consider direct TF regulations for TF to lig</span>

write_csv<span class="token punctuation">(</span>combined_SIF_moon<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/combined_SIF_moon.csv"</span><span class="token punctuation">)</span>
write_csv<span class="token punctuation">(</span>combined_ATT_moon<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/combined_ATT_moon.csv"</span><span class="token punctuation">)</span>

write_csv<span class="token punctuation">(</span>moon_res_rec_to_TFmet<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/moon_res_rec_to_TFmet.csv"</span><span class="token punctuation">)</span>
write_csv<span class="token punctuation">(</span>moon_res_TF_lig<span class="token punctuation">,</span> file <span class="token operator">=</span> <span class="token string">"results/cosmos/moon/moon_res_TF_lig.csv"</span><span class="token punctuation">)</span>
</code></pre><h3 id="further-analyses">Further analyses</h3><p>Further MOFA-COSMOS downstream analyses focusing on analyzing network
control structures and cell line specific MOFA transcriptomics are given
in ./pathway_control_analysis_moon.Rmd and ./Cell_line_MOFA_space.Rmd</p><h3 id="session-info">Session info</h3><pre class="language-r" tabindex="0"><code class="language-r">sessionInfo<span class="token punctuation">(</span><span class="token punctuation">)</span>
</code></pre><pre class="language-bash" tabindex="0"><code class="language-bash"><span class="token comment">## R version 4.2.0 (2022-04-22)</span>
<span class="token comment">## Platform: aarch64-apple-darwin20 (64-bit)</span>
<span class="token comment">## Running under: macOS Monterey 12.6</span>
<span class="token comment">## </span>
<span class="token comment">## Matrix products: default</span>
<span class="token comment">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib</span>
<span class="token comment">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib</span>
<span class="token comment">## </span>
<span class="token comment">## locale:</span>
<span class="token comment">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="token comment">## </span>
<span class="token comment">## attached base packages:</span>
<span class="token comment">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="token comment">## [8] base     </span>
<span class="token comment">## </span>
<span class="token comment">## other attached packages:</span>
<span class="token comment">##  [1] RColorBrewer_1.1-3   gridExtra_2.3        pheatmap_1.0.12     </span>
<span class="token comment">##  [4] ggfortify_0.4.15     ggplot2_3.4.0        tidyr_1.3.0         </span>
<span class="token comment">##  [7] GSEABase_1.58.0      graph_1.74.0         annotate_1.74.0     </span>
<span class="token comment">## [10] XML_3.99-0.13        AnnotationDbi_1.58.0 IRanges_2.30.1      </span>
<span class="token comment">## [13] S4Vectors_0.34.0     Biobase_2.56.0       BiocGenerics_0.42.0 </span>
<span class="token comment">## [16] reshape2_1.4.4       dplyr_1.1.3          decoupleR_2.9.1     </span>
<span class="token comment">## [19] liana_0.1.5          MOFA2_1.6.0          cosmosR_1.5.2       </span>
<span class="token comment">## [22] readr_2.1.4         </span>
<span class="token comment">## </span>
<span class="token comment">## loaded via a namespace (and not attached):</span>
<span class="token comment">##   [1] utf8_1.2.3                  reticulate_1.28            </span>
<span class="token comment">##   [3] tidyselect_1.2.0            RSQLite_2.2.20             </span>
<span class="token comment">##   [5] htmlwidgets_1.6.1           grid_4.2.0                 </span>
<span class="token comment">##   [7] BiocParallel_1.30.4         Rtsne_0.16                 </span>
<span class="token comment">##   [9] devtools_2.4.5              munsell_0.5.0              </span>
<span class="token comment">##  [11] ScaledMatrix_1.4.1          ragg_1.2.5                 </span>
<span class="token comment">##  [13] codetools_0.2-18            statmod_1.5.0              </span>
<span class="token comment">##  [15] scran_1.24.1                future_1.30.0              </span>
<span class="token comment">##  [17] miniUI_0.1.1.1              withr_2.5.0                </span>
<span class="token comment">##  [19] colorspace_2.1-0            progressr_0.13.0           </span>
<span class="token comment">##  [21] filelock_1.0.2              highr_0.10                 </span>
<span class="token comment">##  [23] logger_0.2.2                knitr_1.42                 </span>
<span class="token comment">##  [25] rstudioapi_0.14             SingleCellExperiment_1.18.1</span>
<span class="token comment">##  [27] listenv_0.9.0               labeling_0.4.2             </span>
<span class="token comment">##  [29] MatrixGenerics_1.8.1        GenomeInfoDbData_1.2.8     </span>
<span class="token comment">##  [31] farver_2.1.1                bit64_4.0.5                </span>
<span class="token comment">##  [33] rhdf5_2.40.0                basilisk_1.8.1             </span>
<span class="token comment">##  [35] parallelly_1.34.0           vctrs_0.6.1                </span>
<span class="token comment">##  [37] generics_0.1.3              xfun_0.42                  </span>
<span class="token comment">##  [39] R6_2.5.1                    doParallel_1.0.17          </span>
<span class="token comment">##  [41] GenomeInfoDb_1.32.4         clue_0.3-63                </span>
<span class="token comment">##  [43] rsvd_1.0.5                  locfit_1.5-9.7             </span>
<span class="token comment">##  [45] bitops_1.0-7                rhdf5filters_1.8.0         </span>
<span class="token comment">##  [47] cachem_1.0.7                DelayedArray_0.22.0        </span>
<span class="token comment">##  [49] vroom_1.6.1                 promises_1.2.0.1           </span>
<span class="token comment">##  [51] scales_1.2.1                gtable_0.3.1               </span>
<span class="token comment">##  [53] beachmat_2.12.0             globals_0.16.2             </span>
<span class="token comment">##  [55] processx_3.8.0              rlang_1.1.0                </span>
<span class="token comment">##  [57] systemfonts_1.0.4           splines_4.2.0              </span>
<span class="token comment">##  [59] GlobalOptions_0.1.2         checkmate_2.1.0            </span>
<span class="token comment">##  [61] BiocManager_1.30.19         yaml_2.3.7                 </span>
<span class="token comment">##  [63] backports_1.4.1             httpuv_1.6.8               </span>
<span class="token comment">##  [65] tools_4.2.0                 usethis_2.1.6              </span>
<span class="token comment">##  [67] ellipsis_0.3.2              sessioninfo_1.2.2          </span>
<span class="token comment">##  [69] Rcpp_1.0.10                 plyr_1.8.8                 </span>
<span class="token comment">##  [71] sparseMatrixStats_1.8.0     progress_1.2.2             </span>
<span class="token comment">##  [73] zlibbioc_1.42.0             purrr_1.0.1                </span>
<span class="token comment">##  [75] RCurl_1.98-1.10             ps_1.7.2                   </span>
<span class="token comment">##  [77] basilisk.utils_1.8.0        prettyunits_1.1.1          </span>
<span class="token comment">##  [79] GetoptLong_1.0.5            cowplot_1.1.1              </span>
<span class="token comment">##  [81] urlchecker_1.0.1            SeuratObject_4.1.3         </span>
<span class="token comment">##  [83] SummarizedExperiment_1.26.1 ggrepel_0.9.2              </span>
<span class="token comment">##  [85] cluster_2.1.4               fs_1.6.1                   </span>
<span class="token comment">##  [87] magrittr_2.0.3              circlize_0.4.15            </span>
<span class="token comment">##  [89] matrixStats_0.63.0          pkgload_1.3.2              </span>
<span class="token comment">##  [91] hms_1.1.3                   mime_0.12                  </span>
<span class="token comment">##  [93] evaluate_0.20               xtable_1.8-4               </span>
<span class="token comment">##  [95] readxl_1.4.2                shape_1.4.6                </span>
<span class="token comment">##  [97] compiler_4.2.0              tibble_3.2.1               </span>
<span class="token comment">##  [99] crayon_1.5.2                htmltools_0.5.5            </span>
<span class="token comment">## [101] mgcv_1.8-41                 later_1.3.0                </span>
<span class="token comment">## [103] tzdb_0.3.0                  DBI_1.1.3                  </span>
<span class="token comment">## [105] corrplot_0.92               ComplexHeatmap_2.15.4      </span>
<span class="token comment">## [107] rappdirs_0.3.3              Matrix_1.5-3               </span>
<span class="token comment">## [109] cli_3.6.1                   parallel_4.2.0             </span>
<span class="token comment">## [111] metapod_1.4.0               igraph_1.4.2               </span>
<span class="token comment">## [113] GenomicRanges_1.48.0        forcats_1.0.0              </span>
<span class="token comment">## [115] pkgconfig_2.0.3             dir.expiry_1.4.0           </span>
<span class="token comment">## [117] OmnipathR_3.9.6             sp_1.6-0                   </span>
<span class="token comment">## [119] scuttle_1.6.3               xml2_1.3.3                 </span>
<span class="token comment">## [121] foreach_1.5.2               dqrng_0.3.0                </span>
<span class="token comment">## [123] XVector_0.36.0              rvest_1.0.3                </span>
<span class="token comment">## [125] stringr_1.5.0               callr_3.7.3                </span>
<span class="token comment">## [127] digest_0.6.31               Biostrings_2.64.1          </span>
<span class="token comment">## [129] rmarkdown_2.21              cellranger_1.1.0           </span>
<span class="token comment">## [131] uwot_0.1.14                 edgeR_3.38.4               </span>
<span class="token comment">## [133] DelayedMatrixStats_1.18.2   curl_5.0.0                 </span>
<span class="token comment">## [135] shiny_1.7.4                 rjson_0.2.21               </span>
<span class="token comment">## [137] nlme_3.1-161                lifecycle_1.0.3            </span>
<span class="token comment">## [139] jsonlite_1.8.4              Rhdf5lib_1.18.2            </span>
<span class="token comment">## [141] BiocNeighbors_1.14.0        limma_3.52.4               </span>
<span class="token comment">## [143] fansi_1.0.4                 pillar_1.9.0               </span>
<span class="token comment">## [145] lattice_0.20-45             KEGGREST_1.36.3            </span>
<span class="token comment">## [147] fastmap_1.1.1               httr_1.4.5                 </span>
<span class="token comment">## [149] pkgbuild_1.4.0              glue_1.6.2                 </span>
<span class="token comment">## [151] remotes_2.4.2               png_0.1-8                  </span>
<span class="token comment">## [153] iterators_1.0.14            bit_4.0.5                  </span>
<span class="token comment">## [155] bluster_1.6.0               stringi_1.7.12             </span>
<span class="token comment">## [157] profvis_0.3.7               HDF5Array_1.24.2           </span>
<span class="token comment">## [159] blob_1.2.3                  textshaping_0.3.6          </span>
<span class="token comment">## [161] BiocSingular_1.12.0         memoise_2.0.1              </span>
<span class="token comment">## [163] irlba_2.3.5.1               future.apply_1.10.0</span>
</code></pre></div></template></zero-md></div>


<script>
    function getQuery(key) {
        var query = window.location.search.substring(1);
        var key_values = query.split("&");
        var params = {};
        key_values.map(function (key_val) {
            var key_val_arr = key_val.split("=");
            params[key_val_arr[0]] = key_val_arr[1];
        });
        if (typeof params[key] != "undefined") {
            return params[key];
        }
        return "";
    }

    window.onload = function () {
        md = document.createElement("zero-md")
        md.setAttribute("src", getQuery("src"))
        md.setAttribute("no-shadow", "")
        document.getElementById("mdcontainer").append(md)
    }
</script>

</body></html>